{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Genomics_Workshop_part_1.ipynb",
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.10"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/ternithinator/test2/blob/main/Genomics_Workshop_part_1.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "G4yBrceuFbf3"
      },
      "source": [
        "<video src=\"https://raw.githubusercontent.com/PoODL-CES/PoODL_NGS_workshop.io/main/Nature%20of%20Science.mp4\" controls autoplay loop width=\"400\" align=\"right\"></video>\n",
        "\n",
        "## Genomics Learning Workshop: NGS Analysis Pipeline\n",
        "\n",
        "Easy-to-follow workflow for processing Illumina whole-genome resequencing reads for population genomics. Steps include trimming, mapping, sorting, variant calling, variant filtering, PCA, and ADMIXTURE analysis. For more details, see the bottom of this notebook and visit the GitHub repository.\n",
        "Repository link: [Genomics Learning Workshop](https://github.com/PoODL-CES/Genomics_learning_workshop)\n",
        "\n",
        "This workshop introduces:\n",
        "- Handling raw FASTQ files\n",
        "- Performing quality control and trimming\n",
        "- Mapping reads to a reference genome\n",
        "- Calling and filtering variants\n",
        "- Running PCA and ADMIXTURE for population analysis\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "f6704e35",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4bab6cfa-66a9-4dcb-aa88-d10d3c82fc27"
      },
      "source": [
        "# Miniconda installation and environment setup for Colab NGS Workshop\n",
        "\n",
        "# Download and install Miniconda (skip if already installed)\n",
        "!wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh\n",
        "!bash miniconda.sh -b -p /usr/local/miniconda\n",
        "\n",
        "import sys, os\n",
        "sys.path.append('/usr/local/miniconda/lib/python3.8/site-packages')\n",
        "os.environ['PATH'] = \"/usr/local/miniconda/bin:\" + os.environ['PATH']\n",
        "\n",
        "# Explicitly clear potentially problematic environment variables from Python's os.environ\n",
        "if 'CONDA_PREFIX' in os.environ:\n",
        "    del os.environ['CONDA_PREFIX']\n",
        "if 'CONDA_ENVS_PATH' in os.environ:\n",
        "    del os.environ['CONDA_ENVS_PATH']\n",
        "\n",
        "# Accept ToS for main and R conda channels\n",
        "!conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/main\n",
        "!conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/r\n",
        "\n",
        "# Create new conda environment for your workshop if it doesn't exist\n",
        "!if ! conda info --envs | grep -w 'workshop_ngs'; then \\\n",
        "    conda create -y -n workshop_ngs python=3.7; \\\n",
        "else \\\n",
        "    echo \"Environment 'workshop_ngs' already exists, skipping creation.\"; \\\n",
        "fi\n",
        "\n",
        "# Install necessary bioinformatics tools into the environment\n",
        "!conda install -y -n workshop_ngs -c bioconda -c conda-forge trim-galore samtools fastqc bwa gatk4=4.3.0.0\n",
        "\n",
        "# Verify tool versions inside activated environment in one shell session\n",
        "!bash -c \"source /usr/local/miniconda/bin/activate workshop_ngs && \\\n",
        "          trim_galore --version && samtools --version && fastqc --version && \\\n",
        "          bwa --version && gatk --version\""
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PREFIX=/usr/local/miniconda\n",
            "Unpacking bootstrapper...\n",
            "Unpacking payload...\n",
            "\n",
            "Installing base environment...\n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Miniconda3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Miniconda3: /usr/local/miniconda\n",
            "accepted Terms of Service for \u001b[4;94mhttps://repo.anaconda.com/pkgs/main\u001b[0m\n",
            "accepted Terms of Service for \u001b[4;94mhttps://repo.anaconda.com/pkgs/r\u001b[0m\n",
            "\u001b[1;33mJupyter detected\u001b[0m\u001b[1;33m...\u001b[0m\n",
            "\u001b[1;32m2\u001b[0m\u001b[1;32m channel Terms of Service accepted\u001b[0m\n",
            "Retrieving notices: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Channels:\n",
            " - defaults\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 25.9.1\n",
            "    latest version: 25.11.0\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/miniconda/envs/workshop_ngs\n",
            "\n",
            "  added / updated specs:\n",
            "    - python=3.7\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    ca-certificates-2025.12.2  |       h06a4308_0         125 KB\n",
            "    certifi-2022.12.7          |   py37h06a4308_0         150 KB\n",
            "    openssl-1.1.1w             |       h7f8727e_0         3.7 MB\n",
            "    pip-22.3.1                 |   py37h06a4308_0         2.7 MB\n",
            "    python-3.7.16              |       h7a1cb2a_0        44.8 MB\n",
            "    setuptools-65.6.3          |   py37h06a4308_0         1.1 MB\n",
            "    wheel-0.38.4               |   py37h06a4308_0          63 KB\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:        52.7 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main \n",
            "  _openmp_mutex      pkgs/main/linux-64::_openmp_mutex-5.1-1_gnu \n",
            "  ca-certificates    pkgs/main/linux-64::ca-certificates-2025.12.2-h06a4308_0 \n",
            "  certifi            pkgs/main/linux-64::certifi-2022.12.7-py37h06a4308_0 \n",
            "  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.44-h153f514_2 \n",
            "  libffi             pkgs/main/linux-64::libffi-3.4.4-h6a678d5_1 \n",
            "  libgcc             pkgs/main/linux-64::libgcc-15.2.0-h69a1729_7 \n",
            "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-15.2.0-h166f726_7 \n",
            "  libgomp            pkgs/main/linux-64::libgomp-15.2.0-h4751f2c_7 \n",
            "  libstdcxx          pkgs/main/linux-64::libstdcxx-15.2.0-h39759b7_7 \n",
            "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-15.2.0-hc03a8fd_7 \n",
            "  libxcb             pkgs/main/linux-64::libxcb-1.17.0-h9b100fa_0 \n",
            "  libzlib            pkgs/main/linux-64::libzlib-1.3.1-hb25bd0a_0 \n",
            "  ncurses            pkgs/main/linux-64::ncurses-6.5-h7934f7d_0 \n",
            "  openssl            pkgs/main/linux-64::openssl-1.1.1w-h7f8727e_0 \n",
            "  pip                pkgs/main/linux-64::pip-22.3.1-py37h06a4308_0 \n",
            "  pthread-stubs      pkgs/main/linux-64::pthread-stubs-0.3-h0ce48e5_1 \n",
            "  python             pkgs/main/linux-64::python-3.7.16-h7a1cb2a_0 \n",
            "  readline           pkgs/main/linux-64::readline-8.3-hc2a1206_0 \n",
            "  setuptools         pkgs/main/linux-64::setuptools-65.6.3-py37h06a4308_0 \n",
            "  sqlite             pkgs/main/linux-64::sqlite-3.51.0-h2a70700_0 \n",
            "  tk                 pkgs/main/linux-64::tk-8.6.15-h54e0aa7_0 \n",
            "  wheel              pkgs/main/linux-64::wheel-0.38.4-py37h06a4308_0 \n",
            "  xorg-libx11        pkgs/main/linux-64::xorg-libx11-1.8.12-h9b100fa_1 \n",
            "  xorg-libxau        pkgs/main/linux-64::xorg-libxau-1.0.12-h9b100fa_0 \n",
            "  xorg-libxdmcp      pkgs/main/linux-64::xorg-libxdmcp-1.1.5-h9b100fa_0 \n",
            "  xorg-xorgproto     pkgs/main/linux-64::xorg-xorgproto-2024.1-h5eee18b_1 \n",
            "  xz                 pkgs/main/linux-64::xz-5.6.4-h5eee18b_1 \n",
            "  zlib               pkgs/main/linux-64::zlib-1.3.1-hb25bd0a_0 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages:\n",
            "python-3.7.16        | 44.8 MB   | :   0% 0/1 [00:00<?, ?it/s]\n",
            "openssl-1.1.1w       | 3.7 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\n",
            "\n",
            "pip-22.3.1           | 2.7 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "setuptools-65.6.3    | 1.1 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "certifi-2022.12.7    | 150 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2025 | 125 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "wheel-0.38.4         | 63 KB     | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "pip-22.3.1           | 2.7 MB    | :   1% 0.00579954747671544/1 [00:00<01:06, 66.77s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "python-3.7.16        | 44.8 MB   | :   0% 0.00034843833210413685/1 [00:00<19:07, 1148.23s/it]\n",
            "\n",
            "\n",
            "\n",
            "certifi-2022.12.7    | 150 KB    | :  11% 0.10679390158848107/1 [00:00<00:03,  3.66s/it]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openssl-1.1.1w       | 3.7 MB    | :   0% 0.004187193943698291/1 [00:00<01:39, 99.60s/it]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "certifi-2022.12.7    | 150 KB    | : 100% 1.0/1 [00:00<00:00,  3.66s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2025 | 125 KB    | :  13% 0.1275873347142835/1 [00:00<00:03,  3.75s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "setuptools-65.6.3    | 1.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.64it/s]                \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "setuptools-65.6.3    | 1.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.64it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "python-3.7.16        | 44.8 MB   | :   6% 0.057840763129286724/1 [00:00<00:06,  6.64s/it]    \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2025 | 125 KB    | : 100% 1.0/1 [00:00<00:00,  3.75s/it]               \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openssl-1.1.1w       | 3.7 MB    | :  82% 0.8248772069085634/1 [00:00<00:00,  2.10it/s]  \u001b[A\n",
            "\n",
            "pip-22.3.1           | 2.7 MB    | : 100% 1.0/1 [00:00<00:00,  2.28it/s]                \u001b[A\u001b[A\n",
            "\n",
            "pip-22.3.1           | 2.7 MB    | : 100% 1.0/1 [00:00<00:00,  2.28it/s]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2025 | 125 KB    | : 100% 1.0/1 [00:00<00:00,  3.75s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "wheel-0.38.4         | 63 KB     | :  25% 0.25267963171450164/1 [00:00<00:01,  2.25s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "python-3.7.16        | 44.8 MB   | :  17% 0.1717800977273395/1 [00:00<00:01,  2.41s/it]  \n",
            "python-3.7.16        | 44.8 MB   | :  94% 0.939389743352753/1 [00:01<00:00,  1.03it/s] \n",
            "\n",
            "\n",
            "setuptools-65.6.3    | 1.1 MB    | : 100% 1.0/1 [00:01<00:00,  2.64it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "wheel-0.38.4         | 63 KB     | : 100% 1.0/1 [00:01<00:00,  1.30s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "wheel-0.38.4         | 63 KB     | : 100% 1.0/1 [00:01<00:00,  1.30s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openssl-1.1.1w       | 3.7 MB    | : 100% 1.0/1 [00:01<00:00,  2.10it/s]\u001b[A\n",
            "\n",
            "                                                                        \n",
            "                                                                        \u001b[A\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "Preparing transaction: | \b\b/ \b\b- \b\bdone\n",
            "Verifying transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "Executing transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate workshop_ngs\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n",
            "\u001b[1;33mJupyter detected\u001b[0m\u001b[1;33m...\u001b[0m\n",
            "\u001b[1;32m2\u001b[0m\u001b[1;32m channel Terms of Service accepted\u001b[0m\n",
            "Channels:\n",
            " - bioconda\n",
            " - conda-forge\n",
            " - defaults\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Solving environment: | \b\b/ \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 25.9.1\n",
            "    latest version: 25.11.0\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/miniconda/envs/workshop_ngs\n",
            "\n",
            "  added / updated specs:\n",
            "    - bwa\n",
            "    - fastqc\n",
            "    - gatk4=4.3.0.0\n",
            "    - samtools\n",
            "    - trim-galore\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    alsa-lib-1.2.14            |       hb9d3cd8_0         553 KB  conda-forge\n",
            "    bwa-0.7.19                 |       h577a1d6_1         639 KB  bioconda\n",
            "    bzip2-1.0.8                |       hda65f42_8         254 KB  conda-forge\n",
            "    c-ares-1.34.6              |       hb03c661_0         203 KB  conda-forge\n",
            "    cairo-1.18.4               |       h3394656_0         955 KB  conda-forge\n",
            "    certifi-2024.8.30          |     pyhd8ed1ab_0         160 KB  conda-forge\n",
            "    cutadapt-4.4               |   py37h8902056_0         263 KB  bioconda\n",
            "    dnaio-0.10.0               |   py37h8902056_1          90 KB  bioconda\n",
            "    fastqc-0.12.1              |       hdfd78af_0        11.1 MB  bioconda\n",
            "    font-ttf-dejavu-sans-mono-2.37|       hab24e00_0         388 KB  conda-forge\n",
            "    font-ttf-inconsolata-3.000 |       h77eed37_0          94 KB  conda-forge\n",
            "    font-ttf-source-code-pro-2.038|       h77eed37_0         684 KB  conda-forge\n",
            "    font-ttf-ubuntu-0.83       |       h77eed37_3         1.5 MB  conda-forge\n",
            "    fontconfig-2.15.0          |       h7e30c49_1         259 KB  conda-forge\n",
            "    fonts-conda-ecosystem-1    |                0           4 KB  conda-forge\n",
            "    fonts-conda-forge-1        |       hc364b38_1           4 KB  conda-forge\n",
            "    freetype-2.14.1            |       ha770c72_0         169 KB  conda-forge\n",
            "    gatk4-4.3.0.0              |   py36hdfd78af_0       280.5 MB  bioconda\n",
            "    giflib-5.2.2               |       hd590300_0          75 KB  conda-forge\n",
            "    graphite2-1.3.14           |       hecca717_2          97 KB  conda-forge\n",
            "    harfbuzz-11.4.5            |       h15599e2_0         2.3 MB  conda-forge\n",
            "    htslib-1.21                |       h5efdd21_0         3.0 MB  bioconda\n",
            "    icu-75.1                   |       he02047a_0        11.6 MB  conda-forge\n",
            "    isa-l-2.31.1               |       hb9d3cd8_1         154 KB  conda-forge\n",
            "    keyutils-1.6.3             |       hb9d3cd8_0         131 KB  conda-forge\n",
            "    krb5-1.21.3                |       h659f571_0         1.3 MB  conda-forge\n",
            "    lcms2-2.17                 |       h717163a_0         242 KB  conda-forge\n",
            "    lerc-4.0.0                 |       h0aef613_1         258 KB  conda-forge\n",
            "    libcups-2.3.3              |       hb8b1518_5         4.3 MB  conda-forge\n",
            "    libcurl-8.17.0             |       h4e3cde8_1         449 KB  conda-forge\n",
            "    libdeflate-1.22            |       hb9d3cd8_0          71 KB  conda-forge\n",
            "    libedit-3.1.20250104       | pl5321h7949ede_0         132 KB  conda-forge\n",
            "    libev-4.33                 |       hd590300_2         110 KB  conda-forge\n",
            "    libexpat-2.7.3             |       hecca717_0          75 KB  conda-forge\n",
            "    libfreetype-2.14.1         |       ha770c72_0           7 KB  conda-forge\n",
            "    libfreetype6-2.14.1        |       h73754d4_0         378 KB  conda-forge\n",
            "    libglib-2.84.4             |       h77a78f3_0         2.8 MB\n",
            "    libiconv-1.18              |       h3b78370_2         772 KB  conda-forge\n",
            "    libjpeg-turbo-3.1.2        |       hb03c661_0         619 KB  conda-forge\n",
            "    libnghttp2-1.67.0          |       had1ee68_0         651 KB  conda-forge\n",
            "    libnsl-2.0.1               |       hb9d3cd8_1          33 KB  conda-forge\n",
            "    libpng-1.6.53              |       h421ea60_0         310 KB  conda-forge\n",
            "    libssh2-1.11.1             |       hcf80075_0         298 KB  conda-forge\n",
            "    libtiff-4.7.0              |       he137b08_1         418 KB  conda-forge\n",
            "    libuuid-2.41.2             |       h5347b49_1          39 KB  conda-forge\n",
            "    libwebp-base-1.6.0         |       hd42ef1d_0         419 KB  conda-forge\n",
            "    libxcrypt-4.4.36           |       hd590300_1          98 KB  conda-forge\n",
            "    openjdk-11.0.27            |       h5ddf6bc_0       164.0 MB  conda-forge\n",
            "    openssl-3.6.0              |       h26f9b46_0         3.0 MB  conda-forge\n",
            "    pbzip2-1.1.13              |       h1fcc475_2          41 KB  conda-forge\n",
            "    pcre2-10.46                |       h1321c63_0         1.2 MB  conda-forge\n",
            "    perl-5.32.1                | 7_hd590300_perl5        12.7 MB  conda-forge\n",
            "    pigz-2.8                   |       h421ea60_2          79 KB  conda-forge\n",
            "    pixman-0.46.4              |       h54a6638_1         440 KB  conda-forge\n",
            "    python-3.7.12              |hf930737_100_cpython        57.3 MB  conda-forge\n",
            "    python-isal-1.1.0          |   py37h540881e_0          47 KB  conda-forge\n",
            "    python_abi-3.7             |          4_cp37m           6 KB  conda-forge\n",
            "    samtools-1.21              |       h96c455f_1         476 KB  bioconda\n",
            "    trim-galore-0.6.10         |       hdfd78af_1          45 KB  bioconda\n",
            "    xopen-1.6.0                |   py37h89c1867_0          27 KB  conda-forge\n",
            "    xorg-libice-1.1.2          |       hb9d3cd8_0          57 KB  conda-forge\n",
            "    xorg-libsm-1.2.6           |       he73a12e_0          27 KB  conda-forge\n",
            "    xorg-libxext-1.3.6         |       hb9d3cd8_0          49 KB  conda-forge\n",
            "    xorg-libxfixes-6.0.2       |       hb03c661_0          20 KB  conda-forge\n",
            "    xorg-libxi-1.8.2           |       hb9d3cd8_0          46 KB  conda-forge\n",
            "    xorg-libxrandr-1.5.4       |       hb9d3cd8_0          29 KB  conda-forge\n",
            "    xorg-libxrender-0.9.12     |       hb9d3cd8_0          32 KB  conda-forge\n",
            "    xorg-libxt-1.3.1           |       hb9d3cd8_0         371 KB  conda-forge\n",
            "    xorg-libxtst-1.2.5         |       hb9d3cd8_3          32 KB  conda-forge\n",
            "    zstd-1.5.7                 |       hb78ec9c_6         587 KB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:       569.4 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  alsa-lib           conda-forge/linux-64::alsa-lib-1.2.14-hb9d3cd8_0 \n",
            "  bwa                bioconda/linux-64::bwa-0.7.19-h577a1d6_1 \n",
            "  bzip2              conda-forge/linux-64::bzip2-1.0.8-hda65f42_8 \n",
            "  c-ares             conda-forge/linux-64::c-ares-1.34.6-hb03c661_0 \n",
            "  cairo              conda-forge/linux-64::cairo-1.18.4-h3394656_0 \n",
            "  cutadapt           bioconda/linux-64::cutadapt-4.4-py37h8902056_0 \n",
            "  dnaio              bioconda/linux-64::dnaio-0.10.0-py37h8902056_1 \n",
            "  fastqc             bioconda/noarch::fastqc-0.12.1-hdfd78af_0 \n",
            "  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0 \n",
            "  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0 \n",
            "  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0 \n",
            "  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-h77eed37_3 \n",
            "  fontconfig         conda-forge/linux-64::fontconfig-2.15.0-h7e30c49_1 \n",
            "  fonts-conda-ecosy~ conda-forge/noarch::fonts-conda-ecosystem-1-0 \n",
            "  fonts-conda-forge  conda-forge/noarch::fonts-conda-forge-1-hc364b38_1 \n",
            "  freetype           conda-forge/linux-64::freetype-2.14.1-ha770c72_0 \n",
            "  gatk4              bioconda/noarch::gatk4-4.3.0.0-py36hdfd78af_0 \n",
            "  giflib             conda-forge/linux-64::giflib-5.2.2-hd590300_0 \n",
            "  graphite2          conda-forge/linux-64::graphite2-1.3.14-hecca717_2 \n",
            "  harfbuzz           conda-forge/linux-64::harfbuzz-11.4.5-h15599e2_0 \n",
            "  htslib             bioconda/linux-64::htslib-1.21-h5efdd21_0 \n",
            "  icu                conda-forge/linux-64::icu-75.1-he02047a_0 \n",
            "  isa-l              conda-forge/linux-64::isa-l-2.31.1-hb9d3cd8_1 \n",
            "  keyutils           conda-forge/linux-64::keyutils-1.6.3-hb9d3cd8_0 \n",
            "  krb5               conda-forge/linux-64::krb5-1.21.3-h659f571_0 \n",
            "  lcms2              conda-forge/linux-64::lcms2-2.17-h717163a_0 \n",
            "  lerc               conda-forge/linux-64::lerc-4.0.0-h0aef613_1 \n",
            "  libcups            conda-forge/linux-64::libcups-2.3.3-hb8b1518_5 \n",
            "  libcurl            conda-forge/linux-64::libcurl-8.17.0-h4e3cde8_1 \n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.22-hb9d3cd8_0 \n",
            "  libedit            conda-forge/linux-64::libedit-3.1.20250104-pl5321h7949ede_0 \n",
            "  libev              conda-forge/linux-64::libev-4.33-hd590300_2 \n",
            "  libexpat           conda-forge/linux-64::libexpat-2.7.3-hecca717_0 \n",
            "  libfreetype        conda-forge/linux-64::libfreetype-2.14.1-ha770c72_0 \n",
            "  libfreetype6       conda-forge/linux-64::libfreetype6-2.14.1-h73754d4_0 \n",
            "  libglib            pkgs/main/linux-64::libglib-2.84.4-h77a78f3_0 \n",
            "  libiconv           conda-forge/linux-64::libiconv-1.18-h3b78370_2 \n",
            "  libjpeg-turbo      conda-forge/linux-64::libjpeg-turbo-3.1.2-hb03c661_0 \n",
            "  libnghttp2         conda-forge/linux-64::libnghttp2-1.67.0-had1ee68_0 \n",
            "  libnsl             conda-forge/linux-64::libnsl-2.0.1-hb9d3cd8_1 \n",
            "  libpng             conda-forge/linux-64::libpng-1.6.53-h421ea60_0 \n",
            "  libssh2            conda-forge/linux-64::libssh2-1.11.1-hcf80075_0 \n",
            "  libtiff            conda-forge/linux-64::libtiff-4.7.0-he137b08_1 \n",
            "  libuuid            conda-forge/linux-64::libuuid-2.41.2-h5347b49_1 \n",
            "  libwebp-base       conda-forge/linux-64::libwebp-base-1.6.0-hd42ef1d_0 \n",
            "  libxcrypt          conda-forge/linux-64::libxcrypt-4.4.36-hd590300_1 \n",
            "  openjdk            conda-forge/linux-64::openjdk-11.0.27-h5ddf6bc_0 \n",
            "  pbzip2             conda-forge/linux-64::pbzip2-1.1.13-h1fcc475_2 \n",
            "  pcre2              conda-forge/linux-64::pcre2-10.46-h1321c63_0 \n",
            "  perl               conda-forge/linux-64::perl-5.32.1-7_hd590300_perl5 \n",
            "  pigz               conda-forge/linux-64::pigz-2.8-h421ea60_2 \n",
            "  pixman             conda-forge/linux-64::pixman-0.46.4-h54a6638_1 \n",
            "  python-isal        conda-forge/linux-64::python-isal-1.1.0-py37h540881e_0 \n",
            "  python_abi         conda-forge/linux-64::python_abi-3.7-4_cp37m \n",
            "  samtools           bioconda/linux-64::samtools-1.21-h96c455f_1 \n",
            "  trim-galore        bioconda/noarch::trim-galore-0.6.10-hdfd78af_1 \n",
            "  xopen              conda-forge/linux-64::xopen-1.6.0-py37h89c1867_0 \n",
            "  xorg-libice        conda-forge/linux-64::xorg-libice-1.1.2-hb9d3cd8_0 \n",
            "  xorg-libsm         conda-forge/linux-64::xorg-libsm-1.2.6-he73a12e_0 \n",
            "  xorg-libxext       conda-forge/linux-64::xorg-libxext-1.3.6-hb9d3cd8_0 \n",
            "  xorg-libxfixes     conda-forge/linux-64::xorg-libxfixes-6.0.2-hb03c661_0 \n",
            "  xorg-libxi         conda-forge/linux-64::xorg-libxi-1.8.2-hb9d3cd8_0 \n",
            "  xorg-libxrandr     conda-forge/linux-64::xorg-libxrandr-1.5.4-hb9d3cd8_0 \n",
            "  xorg-libxrender    conda-forge/linux-64::xorg-libxrender-0.9.12-hb9d3cd8_0 \n",
            "  xorg-libxt         conda-forge/linux-64::xorg-libxt-1.3.1-hb9d3cd8_0 \n",
            "  xorg-libxtst       conda-forge/linux-64::xorg-libxtst-1.2.5-hb9d3cd8_3 \n",
            "  zstd               conda-forge/linux-64::zstd-1.5.7-hb78ec9c_6 \n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  certifi            pkgs/main/linux-64::certifi-2022.12.7~ --> conda-forge/noarch::certifi-2024.8.30-pyhd8ed1ab_0 \n",
            "  openssl              pkgs/main::openssl-1.1.1w-h7f8727e_0 --> conda-forge::openssl-3.6.0-h26f9b46_0 \n",
            "\n",
            "The following packages will be SUPERSEDED by a higher-priority channel:\n",
            "\n",
            "  python                pkgs/main::python-3.7.16-h7a1cb2a_0 --> conda-forge::python-3.7.12-hf930737_100_cpython \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages:\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :   0% 0/1 [00:00<?, ?it/s]\n",
            "openjdk-11.0.27      | 164.0 MB  | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "icu-75.1             | 11.6 MB   | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "fastqc-0.12.1        | 11.1 MB   | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libcups-2.3.3        | 4.3 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.21          | 3.0 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.6.0        | 3.0 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libglib-2.84.4       | 2.8 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "harfbuzz-11.4.5      | 2.3 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-ubuntu-0.83 | 1.5 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "krb5-1.21.3          | 1.3 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "pcre2-10.46          | 1.2 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "cairo-1.18.4         | 955 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libiconv-1.18        | 772 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-source-code | 684 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libnghttp2-1.67.0    | 651 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "bwa-0.7.19           | 639 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "icu-75.1             | 11.6 MB   | :   0% 0.00135078949540213/1 [00:00<02:54, 174.38s/it]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :   0% 5.5705534289751753e-05/1 [00:00<1:26:29, 5190.18s/it]\n",
            "openjdk-11.0.27      | 164.0 MB  | :   0% 9.526528731089631e-05/1 [00:00<51:38, 3098.35s/it]\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :   0% 0.00027256019984598284/1 [00:00<18:10, 1090.99s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "icu-75.1             | 11.6 MB   | :  24% 0.23773895119077487/1 [00:00<00:00,  1.12s/it] \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :   1% 0.008021596937724253/1 [00:00<00:37, 37.82s/it]      \n",
            "openjdk-11.0.27      | 164.0 MB  | :   1% 0.013813466660079965/1 [00:00<00:21, 22.27s/it]   \u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :   4% 0.03870354837812957/1 [00:00<00:07,  8.02s/it]     \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "icu-75.1             | 11.6 MB   | :  48% 0.478179481372354/1 [00:00<00:00,  1.42it/s]  \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :   2% 0.018772765055646342/1 [00:00<00:18, 19.09s/it]\n",
            "openjdk-11.0.27      | 164.0 MB  | :   3% 0.032104401823772055/1 [00:00<00:10, 11.28s/it]\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :   9% 0.0907625465487123/1 [00:00<00:03,  4.01s/it] \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "icu-75.1             | 11.6 MB   | :  68% 0.6834994846734778/1 [00:00<00:00,  1.63it/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :   3% 0.03325620397098179/1 [00:00<00:12, 12.50s/it] \n",
            "openjdk-11.0.27      | 164.0 MB  | :   5% 0.047823174230069945/1 [00:00<00:08,  9.09s/it]\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  15% 0.1450020263180629/1 [00:00<00:02,  2.94s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :   5% 0.04701547094055048/1 [00:00<00:09, 10.28s/it]\n",
            "openjdk-11.0.27      | 164.0 MB  | :   7% 0.07459271996443181/1 [00:00<00:05,  6.32s/it] \u001b[A\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :   6% 0.06222308180165271/1 [00:00<00:08,  8.78s/it]\n",
            "openjdk-11.0.27      | 164.0 MB  | :   9% 0.09497949144896363/1 [00:00<00:05,  5.83s/it]\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  25% 0.25157306445784217/1 [00:00<00:01,  2.36s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | : 100% 1.0/1 [00:00<00:00,  1.25it/s]               \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | : 100% 1.0/1 [00:00<00:00,  1.25it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :   8% 0.0762051709083804/1 [00:00<00:07,  8.26s/it] \n",
            "openjdk-11.0.27      | 164.0 MB  | :  11% 0.11498520178425185/1 [00:00<00:04,  5.55s/it]\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  31% 0.3101735074247285/1 [00:00<00:01,  2.16s/it] \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "fastqc-0.12.1        | 11.1 MB   | :   0% 0.0014046288797150209/1 [00:00<10:44, 645.45s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libcups-2.3.3        | 4.3 MB    | :   0% 0.0036218772527583545/1 [00:00<04:17, 258.26s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "fastqc-0.12.1        | 11.1 MB   | :  17% 0.16574620780637245/1 [00:01<00:03,  4.45s/it]   \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  13% 0.1343240551083638/1 [00:01<00:05,  5.86s/it] \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libcups-2.3.3        | 4.3 MB    | :  38% 0.3766752342868689/1 [00:01<00:01,  2.01s/it]    \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :   9% 0.08918456039789255/1 [00:01<00:09, 10.76s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "fastqc-0.12.1        | 11.1 MB   | :  30% 0.3033998380184445/1 [00:01<00:01,  2.43s/it] \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libcups-2.3.3        | 4.3 MB    | :  73% 0.7279973278044293/1 [00:01<00:00,  1.03s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  15% 0.1524244596974341/1 [00:01<00:05,  6.69s/it]\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  40% 0.40311653557220867/1 [00:01<00:01,  2.57s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  10% 0.0999914340501044/1 [00:01<00:10, 11.41s/it] \n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  44% 0.444818246148644/1 [00:01<00:01,  2.55s/it]  \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "fastqc-0.12.1        | 11.1 MB   | :  62% 0.6180367070746091/1 [00:01<00:00,  1.21s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  17% 0.16852429325297558/1 [00:01<00:06,  7.39s/it]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  11% 0.10979560808510071/1 [00:01<00:11, 12.40s/it]\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  49% 0.4862473965252334/1 [00:01<00:01,  2.55s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.21          | 3.0 MB    | :   1% 0.0051733567077445/1 [00:01<04:28, 270.37s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "fastqc-0.12.1        | 11.1 MB   | :  77% 0.7739505127229764/1 [00:01<00:00,  1.00s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  18% 0.18395726979734078/1 [00:01<00:05,  7.16s/it]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.21          | 3.0 MB    | :  68% 0.6777097287145295/1 [00:01<00:00,  1.60s/it] \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  12% 0.11865278803717123/1 [00:01<00:11, 13.00s/it]\n",
            "openjdk-11.0.27      | 164.0 MB  | :  20% 0.1986281240432188/1 [00:01<00:05,  7.15s/it] \u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  53% 0.5271314265021309/1 [00:01<00:01,  2.94s/it]\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  22% 0.21539481460993656/1 [00:01<00:05,  6.79s/it]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  13% 0.12684150157776475/1 [00:01<00:11, 13.48s/it]\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  57% 0.5674703360793363/1 [00:01<00:01,  2.81s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  14% 0.13753696416139707/1 [00:01<00:10, 12.17s/it]\n",
            "openjdk-11.0.27      | 164.0 MB  | :  23% 0.23063726057967998/1 [00:01<00:05,  7.26s/it]\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  61% 0.6108074078548476/1 [00:01<00:01,  2.67s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  15% 0.14711831605923437/1 [00:01<00:09, 11.69s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "fastqc-0.12.1        | 11.1 MB   | : 100% 1.0/1 [00:01<00:00,  1.29it/s]               \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  66% 0.6585054428278946/1 [00:01<00:00,  2.49s/it]\u001b[A\u001b[A\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  16% 0.15603120154559466/1 [00:01<00:10, 12.61s/it]\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  70% 0.7002071534043299/1 [00:01<00:00,  2.55s/it]\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  26% 0.2613126830937886/1 [00:02<00:05,  7.22s/it] \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.6.0        | 3.0 MB    | : 100% 1.0/1 [00:02<00:00,  1.81s/it]               \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  17% 0.166949486266386/1 [00:02<00:09, 11.66s/it]  \n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  74% 0.7435442251798412/1 [00:02<00:00,  2.52s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "harfbuzz-11.4.5      | 2.3 MB    | :   1% 0.00681973894851813/1 [00:02<05:03, 305.63s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  18% 0.1775892433157286/1 [00:02<00:09, 10.97s/it]\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  78% 0.7847008153565846/1 [00:02<00:00,  2.50s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "harfbuzz-11.4.5      | 2.3 MB    | :  91% 0.9138450191014295/1 [00:02<00:00,  1.71s/it]  \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libglib-2.84.4       | 2.8 MB    | :   1% 0.00551151197552921/1 [00:02<06:37, 399.91s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "harfbuzz-11.4.5      | 2.3 MB    | : 100% 1.0/1 [00:02<00:00,  1.71s/it]               \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  31% 0.30570630698066625/1 [00:02<00:04,  7.14s/it]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-ubuntu-0.83 | 1.5 MB    | :   1% 0.010110434778315882/1 [00:02<03:45, 227.56s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libglib-2.84.4       | 2.8 MB    | :  50% 0.501547589773158/1 [00:02<00:01,  3.29s/it]   \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  19% 0.18700347861069663/1 [00:02<00:10, 12.38s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-ubuntu-0.83 | 1.5 MB    | : 100% 1.0/1 [00:02<00:00,  1.73s/it]                  \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-ubuntu-0.83 | 1.5 MB    | : 100% 1.0/1 [00:02<00:00,  1.73s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  20% 0.19597206963134667/1 [00:02<00:09, 12.05s/it]\n",
            "openjdk-11.0.27      | 164.0 MB  | :  32% 0.3199008347899898/1 [00:02<00:05,  8.01s/it] \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "krb5-1.21.3          | 1.3 MB    | :   1% 0.011958923317345767/1 [00:02<03:23, 205.63s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  33% 0.33276164857696083/1 [00:02<00:05,  8.17s/it]\u001b[A\n",
            "\n",
            "python-3.7.12        | 57.3 MB   | :  90% 0.8991760992918975/1 [00:02<00:00,  2.88s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libglib-2.84.4       | 2.8 MB    | : 100% 1.0/1 [00:02<00:00,  1.63s/it]              \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libglib-2.84.4       | 2.8 MB    | : 100% 1.0/1 [00:02<00:00,  1.63s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "krb5-1.21.3          | 1.3 MB    | : 100% 1.0/1 [00:02<00:00,  1.83s/it]                  \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  20% 0.20460642744625818/1 [00:02<00:10, 13.64s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "pcre2-10.46          | 1.2 MB    | :   1% 0.013549711911490212/1 [00:02<03:08, 191.57s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "cairo-1.18.4         | 955 KB    | :   2% 0.016750603712859646/1 [00:02<02:33, 155.84s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  35% 0.34743250282283883/1 [00:02<00:05,  7.80s/it]\u001b[A\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  21% 0.2129065520554312/1 [00:02<00:10, 13.20s/it] \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "cairo-1.18.4         | 955 KB    | : 100% 1.0/1 [00:02<00:00, 155.84s/it]                 \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "pcre2-10.46          | 1.2 MB    | : 100% 1.0/1 [00:02<00:00,  1.96s/it]                  \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "pcre2-10.46          | 1.2 MB    | : 100% 1.0/1 [00:02<00:00,  1.96s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  36% 0.36048384718443166/1 [00:02<00:05,  8.22s/it]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libiconv-1.18        | 772 KB    | :   2% 0.02073462114769368/1 [00:02<02:11, 133.89s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  22% 0.22081673792457596/1 [00:02<00:10, 13.12s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-source-code | 684 KB    | :   2% 0.02337852839697837/1 [00:02<01:56, 119.04s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libiconv-1.18        | 772 KB    | : 100% 1.0/1 [00:02<00:00, 133.89s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-source-code | 684 KB    | : 100% 1.0/1 [00:02<00:00, 119.04s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  23% 0.22867121825943093/1 [00:02<00:10, 13.66s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libnghttp2-1.67.0    | 651 KB    | :   2% 0.02457845784578458/1 [00:02<01:55, 118.10s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "bwa-0.7.19           | 639 KB    | :   3% 0.025042108770546176/1 [00:02<01:53, 116.33s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "bwa-0.7.19           | 639 KB    | : 100% 1.0/1 [00:02<00:00, 116.33s/it]                 \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libnghttp2-1.67.0    | 651 KB    | : 100% 1.0/1 [00:02<00:00, 118.10s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | :  39% 0.3896350251015659/1 [00:02<00:04,  7.55s/it]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  24% 0.23619146538854743/1 [00:03<00:11, 14.67s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  25% 0.24644128369786175/1 [00:03<00:09, 12.97s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  26% 0.25501993597848355/1 [00:03<00:09, 12.64s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  26% 0.26487981554776957/1 [00:03<00:08, 11.86s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  28% 0.2767450943514867/1 [00:03<00:07, 10.87s/it] \n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  29% 0.2878304956751473/1 [00:03<00:07, 10.33s/it]\n",
            "\n",
            "\n",
            "\n",
            "icu-75.1             | 11.6 MB   | : 100% 1.0/1 [00:03<00:00,  1.92it/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  30% 0.29763466971014363/1 [00:03<00:07, 10.45s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  32% 0.32114240518041887/1 [00:03<00:06,  9.42s/it]\n",
            "openjdk-11.0.27      | 164.0 MB  | :  52% 0.5198626728555612/1 [00:03<00:03,  6.64s/it]\u001b[A\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  33% 0.33183786776405116/1 [00:03<00:06,  9.79s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  34% 0.3421433916076553/1 [00:04<00:06, 10.16s/it] \n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  35% 0.3520589767112311/1 [00:04<00:06, 10.23s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libcups-2.3.3        | 4.3 MB    | : 100% 1.0/1 [00:04<00:00,  1.03s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  36% 0.3619188562805171/1 [00:04<00:07, 11.74s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  37% 0.37077603623258765/1 [00:04<00:08, 13.29s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  38% 0.38102585454190196/1 [00:04<00:07, 12.24s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  39% 0.3936153052913859/1 [00:04<00:06, 11.08s/it] \n",
            "\n",
            "python-3.7.12        | 57.3 MB   | : 100% 1.0/1 [00:04<00:00,  2.91s/it]               \u001b[A\u001b[A\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  40% 0.40392082913498994/1 [00:04<00:06, 10.94s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  42% 0.41729015736453035/1 [00:04<00:05,  9.91s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  43% 0.43288770696566087/1 [00:05<00:04,  8.67s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  45% 0.44653556286665/1 [00:05<00:04,  8.32s/it]   \n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  46% 0.4612418239191445/1 [00:05<00:04,  7.89s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  47% 0.47410980234007716/1 [00:05<00:04,  8.13s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  49% 0.4914899290384797/1 [00:05<00:03,  7.39s/it] \n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  51% 0.5068646565024512/1 [00:05<00:03,  7.14s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  52% 0.5240776665979845/1 [00:05<00:03,  6.71s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  54% 0.5391181608562174/1 [00:05<00:03,  6.74s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  57% 0.5684749774269167/1 [00:06<00:03,  8.10s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  58% 0.5846295823709446/1 [00:06<00:03,  7.54s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  60% 0.5997257821634674/1 [00:06<00:02,  7.31s/it]\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  88% 0.8784762757493851/1 [00:08<00:00,  5.58s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | :  99% 0.9886061170402244/1 [00:08<00:00,  4.60s/it]\n",
            "\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | : 100% 1.0/1 [00:09<00:00,  1.25it/s]\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | : 100% 1.0/1 [00:09<00:00,  4.37s/it]              \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.6.0        | 3.0 MB    | : 100% 1.0/1 [00:09<00:00,  1.81s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "harfbuzz-11.4.5      | 2.3 MB    | : 100% 1.0/1 [00:09<00:00,  1.71s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-ubuntu-0.83 | 1.5 MB    | : 100% 1.0/1 [00:09<00:00,  1.73s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libglib-2.84.4       | 2.8 MB    | : 100% 1.0/1 [00:10<00:00,  1.63s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "krb5-1.21.3          | 1.3 MB    | : 100% 1.0/1 [00:10<00:00,  1.83s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "cairo-1.18.4         | 955 KB    | : 100% 1.0/1 [00:10<00:00,  9.70s/it] \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "cairo-1.18.4         | 955 KB    | : 100% 1.0/1 [00:10<00:00,  9.70s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "pcre2-10.46          | 1.2 MB    | : 100% 1.0/1 [00:10<00:00,  1.96s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libiconv-1.18        | 772 KB    | : 100% 1.0/1 [00:10<00:00, 10.05s/it] \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libiconv-1.18        | 772 KB    | : 100% 1.0/1 [00:10<00:00, 10.05s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-source-code | 684 KB    | : 100% 1.0/1 [00:11<00:00, 10.32s/it] \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "font-ttf-source-code | 684 KB    | : 100% 1.0/1 [00:11<00:00, 10.32s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "bwa-0.7.19           | 639 KB    | : 100% 1.0/1 [00:11<00:00, 10.47s/it] \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "bwa-0.7.19           | 639 KB    | : 100% 1.0/1 [00:11<00:00, 10.47s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libnghttp2-1.67.0    | 651 KB    | : 100% 1.0/1 [00:11<00:00, 10.52s/it] \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | : 100% 1.0/1 [00:11<00:00,  4.60s/it]               \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            " ... (more hidden) ...\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "fastqc-0.12.1        | 11.1 MB   | : 100% 1.0/1 [00:11<00:00,  1.29it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "openjdk-11.0.27      | 164.0 MB  | : 100% 1.0/1 [00:15<00:00,  4.37s/it]\u001b[A\n",
            "\n",
            "gatk4-4.3.0.0        | 280.5 MB  | : 100% 1.0/1 [01:00<00:00,  4.60s/it]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \n",
            "                                                                        \u001b[A\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\u001b[A\n",
            "\n",
            "\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\u001b[A\n",
            "\n",
            "\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\u001b[A\n",
            "\n",
            "\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "Preparing transaction: \\ \b\b| \b\b/ \b\bdone\n",
            "Verifying transaction: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Executing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "\n",
            "                        Quality-/Adapter-/RRBS-/Speciality-Trimming\n",
            "                                [powered by Cutadapt]\n",
            "                                  version 0.6.10\n",
            "\n",
            "                               Last update: 02 02 2023\n",
            "\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools 1.21\n",
            "Using htslib 1.21\n",
            "Copyright (C) 2024 Genome Research Ltd.\n",
            "\n",
            "Samtools compilation details:\n",
            "    Features:       build=configure curses=yes \n",
            "    CC:             /opt/conda/conda-bld/samtools_1734249775320/_build_env/bin/x86_64-conda-linux-gnu-cc\n",
            "    CPPFLAGS:       -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /usr/local/miniconda/envs/workshop_ngs/include\n",
            "    CFLAGS:         -Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /usr/local/miniconda/envs/workshop_ngs/include -fdebug-prefix-map=/opt/conda/conda-bld/samtools_1734249775320/work=/usr/local/src/conda/samtools-1.21 -fdebug-prefix-map=/usr/local/miniconda/envs/workshop_ngs=/usr/local/src/conda-prefix\n",
            "    LDFLAGS:        -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/usr/local/miniconda/envs/workshop_ngs/lib -Wl,-rpath-link,/usr/local/miniconda/envs/workshop_ngs/lib -L/usr/local/miniconda/envs/workshop_ngs/lib\n",
            "    HTSDIR:         \n",
            "    LIBS:           \n",
            "    CURSES_LIB:     -ltinfow -lncursesw\n",
            "\n",
            "HTSlib compilation details:\n",
            "    Features:       build=configure libcurl=yes S3=yes GCS=yes libdeflate=yes lzma=yes bzip2=yes plugins=yes plugin-path=/usr/local/miniconda/envs/workshop_ngs/libexec/htslib htscodecs=1.6.1\n",
            "    CC:             /opt/conda/conda-bld/htslib_1726172129301/_build_env/bin/x86_64-conda-linux-gnu-cc\n",
            "    CPPFLAGS:       -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /usr/local/miniconda/envs/workshop_ngs/include\n",
            "    CFLAGS:         -Wall -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /usr/local/miniconda/envs/workshop_ngs/include -fdebug-prefix-map=/opt/conda/conda-bld/htslib_1726172129301/work=/usr/local/src/conda/htslib-1.21 -fdebug-prefix-map=/usr/local/miniconda/envs/workshop_ngs=/usr/local/src/conda-prefix -fvisibility=hidden\n",
            "    LDFLAGS:        -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/usr/local/miniconda/envs/workshop_ngs/lib -Wl,-rpath-link,/usr/local/miniconda/envs/workshop_ngs/lib -L/usr/local/miniconda/envs/workshop_ngs/lib -fvisibility=hidden -rdynamic\n",
            "\n",
            "HTSlib URL scheme handlers present:\n",
            "    built-in:\t file, preload, data\n",
            "    libcurl:\t gophers, smtp, wss, smb, rtsp, tftp, pop3, smbs, imaps, pop3s, ws, ftps, ftp, gopher, imap, http, https, sftp, smtps, dict, scp, mqtt, telnet\n",
            "    Amazon S3:\t s3+https, s3, s3+http\n",
            "    S3 Multipart Upload:\t s3w+https, s3w+http, s3w\n",
            "    Google Cloud Storage:\t gs+http, gs+https, gs\n",
            "    crypt4gh-needed:\t crypt4gh\n",
            "    mem:\t mem\n",
            "FastQC v0.12.1\n",
            "[main] unrecognized command '--version'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "70edf711"
      },
      "source": [
        "To activate the `workshop_ngs` environment for specific commands and ensure its tools are used, you can wrap your commands within a `bash -c \"source /usr/local/miniconda/bin/activate workshop_ngs && your_command_here\"` block. This ensures the environment is properly sourced before the command runs. Let's try listing the environment contents and checking `CONDA_PREFIX` within that activated session:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "e2f52e3b"
      },
      "source": [
        "!bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && conda activate workshop_ngs && \\\n",
        "          echo '--- Environment activated ---' && \\\n",
        "          echo 'CONDA_PREFIX in activated env: '$CONDA_PREFIX && \\\n",
        "          echo '--- Listing contents of activated env ---' && \\\n",
        "          conda list\""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5c5c06e3"
      },
      "source": [
        "Well be using real WGS resequencing data for this workshop. The dataset is a small subset for demonstration purposes. We will download the data from (https://zenodo.org/records/14258052) . Step 1 would be to make a parent directory to sort all the data in one place.\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "f0a0fc21",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e16ddc4b-4d1d-4389-d7c8-ad330fdbb707"
      },
      "source": [
        "%%bash\n",
        "\n",
        "mkdir -p fastq_files\n",
        "ls -F"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "fastq_files/\n",
            "miniconda.sh\n",
            "sample_data/\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "58d954b7"
      },
      "source": [
        "Now, let's navigate into the `fastq_files` directory."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "fd7e272d",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "0eb3b79d-ca8a-4267-ece5-7ba65fd76ca3"
      },
      "source": [
        "%%bash\n",
        "\n",
        "cd fastq_files/ && pwd"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/fastq_files\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8d6ca911"
      },
      "source": [
        "Let's organize the download process by first ensuring the `fastq_files` directory exists, then navigating into it to download all the FASTQ files directly to their intended location."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "bb6bbeb7"
      },
      "source": [
        "# Define the list of FASTQ file links\n",
        "fastq_links = [\n",
        "    \"https://zenodo.org/records/14258052/files/BEN_CI16_sub_1.fq.gz\",\n",
        "    \"https://zenodo.org/records/14258052/files/BEN_CI16_sub_2.fq.gz\",\n",
        "    \"https://zenodo.org/records/14258052/files/BEN_NW10_sub_1.fq.gz\",\n",
        "    \"https://zenodo.org/records/14258052/files/BEN_NW10_sub_2.fq.gz\",\n",
        "    \"https://zenodo.org/records/14258052/files/BEN_SI18_sub_1.fq.gz\",\n",
        "    \"https://zenodo.org/records/14258052/files/BEN_SI18_sub_2.fq.gz\",\n",
        "]\n",
        "\n",
        "# Create the directory if it doesn't exist\n",
        "!mkdir -p fastq_files\n",
        "\n",
        "# Clean up any existing fastq files and duplicates before downloading\n",
        "!rm -f fastq_files/*.fq.gz*\n",
        "\n",
        "# Change to the directory and download files\n",
        "for link in fastq_links:\n",
        "    # Use -nc (no-clobber) to avoid re-downloading if file exists and is complete,\n",
        "    # and -P . to save to current directory (fastq_files after cd)\n",
        "    !bash -c \"cd fastq_files && wget -nc -P . -q {link}\"\n",
        "\n",
        "# List the contents of the fastq_files directory to confirm all downloads\n",
        "!ls -F fastq_files/"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "# Define the list of FASTQ links\n",
        "fastq_links=(\n",
        "\"https://zenodo.org/records/14258052/files/BEN_CI16_sub_1.fq.gz\"\n",
        "\"https://zenodo.org/records/14258052/files/BEN_CI16_sub_2.fq.gz\"\n",
        ")\n",
        "#\"https://zenodo.org/records/14258052/files/BEN_NW10_sub_1.fq.gz\"\n",
        "#\"https://zenodo.org/records/14258052/files/BEN_NW10_sub_2.fq.gz\"\n",
        "#\"https://zenodo.org/records/14258052/files/BEN_SI18_sub_1.fq.gz\"\n",
        "#\"https://zenodo.org/records/14258052/files/BEN_SI18_sub_2.fq.gz\"\n",
        "#)\n",
        "\n",
        "# Create directory\n",
        "mkdir -p fastq_files\n",
        "\n",
        "# Clean old files\n",
        "rm -f fastq_files/*.fq.gz*\n",
        "\n",
        "# Download all files\n",
        "cd fastq_files\n",
        "\n",
        "for link in \"${fastq_links[@]}\"; do\n",
        "    wget -nc -q \"$link\"\n",
        "done\n",
        "\n",
        "# List downloaded files\n",
        "ls -F\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7MLLBiYQPhro",
        "outputId": "117af441-d67f-43e3-eb54-5d18f24e392e"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "BEN_CI16_sub_1.fq.gz\n",
            "BEN_CI16_sub_2.fq.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "To visualise the content of the directory 'fastq_files' , to confirm that all the files have been successfully downloaded."
      ],
      "metadata": {
        "id": "NIKJbD1kB49q"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5e74bb29"
      },
      "source": [
        "Once all the fastq files are downloaded, the next would be to run \"FastQC\" tool on the raw reads for the quality assessment step before the downstream analysis part.\n",
        "\n",
        "***Why it's important:*** Poor-quality reads can lead to false variant calls or poor mapping, so quality control is crucial.\n",
        "\n",
        "To do so, first, let's create a directory to store the FastQC reports."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "93be3c96",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "9e313a42-980f-474d-faca-2e6493aeb73f"
      },
      "source": [
        "%%bash\n",
        "%%time\n",
        "\n",
        "mkdir -p fastqc_results\n",
        "ls -F"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "fastqc_results/\n",
            "fastq_files/\n",
            "miniconda.sh\n",
            "reference/\n",
            "reference.tar.gz\n",
            "sample_data/\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "bash: line 1: fg: no job control\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1cf4153e"
      },
      "source": [
        "Now, we will run FastQC on all the FASTQ files within the `workshop_ngs` environment and save the reports to the `fastqc_results` directory."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ddc193b1"
      },
      "source": [
        "%%time\n",
        "!bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && conda activate workshop_ngs && \\\n",
        "          fastqc fastq_files/*.fq.gz -o fastqc_results\"\n",
        "\n",
        "# List the generated FastQC reports\n",
        "!ls -F fastqc_results/"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "# Load conda\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Activate environment\n",
        "conda activate workshop_ngs\n",
        "\n",
        "# Run FastQC on all FASTQ files\n",
        "fastqc fastq_files/*.fq.gz -o fastqc_results\n",
        "\n",
        "# List the generated FastQC reports\n",
        "ls -F fastqc_results/\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "asc3m80ARavQ",
        "outputId": "1e84d14c-b65b-43e1-eba6-a7cd9f5f628c"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "application/gzip\n",
            "application/gzip\n",
            "application/gzip\n",
            "application/gzip\n",
            "application/gzip\n",
            "application/gzip\n",
            "Analysis complete for BEN_CI16_sub_1.fq.gz\n",
            "Analysis complete for BEN_CI16_sub_2.fq.gz\n",
            "Analysis complete for BEN_NW10_sub_1.fq.gz\n",
            "Analysis complete for BEN_NW10_sub_2.fq.gz\n",
            "Analysis complete for BEN_SI18_sub_1.fq.gz\n",
            "Analysis complete for BEN_SI18_sub_2.fq.gz\n",
            "BEN_CI16_sub_1_fastqc.html\n",
            "BEN_CI16_sub_1_fastqc.zip\n",
            "BEN_CI16_sub_2_fastqc.html\n",
            "BEN_CI16_sub_2_fastqc.zip\n",
            "BEN_NW10_sub_1_fastqc.html\n",
            "BEN_NW10_sub_1_fastqc.zip\n",
            "BEN_NW10_sub_2_fastqc.html\n",
            "BEN_NW10_sub_2_fastqc.zip\n",
            "BEN_SI18_sub_1_fastqc.html\n",
            "BEN_SI18_sub_1_fastqc.zip\n",
            "BEN_SI18_sub_2_fastqc.html\n",
            "BEN_SI18_sub_2_fastqc.zip\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Started analysis of BEN_CI16_sub_1.fq.gz\n",
            "Approx 5% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 10% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 15% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 20% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 25% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 30% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 35% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 40% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 45% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 50% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 55% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 60% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 65% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 70% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 75% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 80% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 85% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 90% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 95% complete for BEN_CI16_sub_1.fq.gz\n",
            "Approx 100% complete for BEN_CI16_sub_1.fq.gz\n",
            "Started analysis of BEN_CI16_sub_2.fq.gz\n",
            "Approx 5% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 10% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 15% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 20% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 25% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 30% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 35% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 40% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 45% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 50% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 55% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 60% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 65% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 70% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 75% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 80% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 85% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 90% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 95% complete for BEN_CI16_sub_2.fq.gz\n",
            "Approx 100% complete for BEN_CI16_sub_2.fq.gz\n",
            "Started analysis of BEN_NW10_sub_1.fq.gz\n",
            "Approx 5% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 10% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 15% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 20% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 25% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 30% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 35% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 40% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 45% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 50% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 55% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 60% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 65% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 70% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 75% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 80% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 85% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 90% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 95% complete for BEN_NW10_sub_1.fq.gz\n",
            "Approx 100% complete for BEN_NW10_sub_1.fq.gz\n",
            "Started analysis of BEN_NW10_sub_2.fq.gz\n",
            "Approx 5% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 10% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 15% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 20% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 25% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 30% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 35% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 40% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 45% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 50% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 55% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 60% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 65% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 70% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 75% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 80% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 85% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 90% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 95% complete for BEN_NW10_sub_2.fq.gz\n",
            "Approx 100% complete for BEN_NW10_sub_2.fq.gz\n",
            "Started analysis of BEN_SI18_sub_1.fq.gz\n",
            "Approx 5% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 10% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 15% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 20% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 25% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 30% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 35% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 40% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 45% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 50% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 55% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 60% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 65% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 70% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 75% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 80% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 85% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 90% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 95% complete for BEN_SI18_sub_1.fq.gz\n",
            "Approx 100% complete for BEN_SI18_sub_1.fq.gz\n",
            "Started analysis of BEN_SI18_sub_2.fq.gz\n",
            "Approx 5% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 10% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 15% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 20% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 25% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 30% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 35% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 40% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 45% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 50% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 55% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 60% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 65% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 70% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 75% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 80% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 85% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 90% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 95% complete for BEN_SI18_sub_2.fq.gz\n",
            "Approx 100% complete for BEN_SI18_sub_2.fq.gz\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "The .html files generated by FastQC are standard web pages, which means you can visualize them by opening them in any web browser!\n",
        "\n",
        "Here's how you can access them from Colab:\n",
        " On the left-hand side of your Colab interface, there's usually a file explorer icon (a folder). Click on it to open the file browser. Navigate to the fastqc_results directory. You'll see all the .html files listed there. You can right-click on any .html file and select 'Download' to save it to your local machine, then open it with your preferred web browser.\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "TLmb_MoTEJ-p"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "26726bab"
      },
      "source": [
        "Post qualitty assessment, the next step will be the trimming of the adapters or low-quality bases using tools such as Trim-galore. This step will ensure producing clean FASTQ files, ready for mapping.\n",
        "\n",
        "Step 1: We'll create a directory to store the trimmed FASTQ files."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "06e02af7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "eae15f02-505c-4e3d-a060-c90a704ecc3d"
      },
      "source": [
        "%%bash\n",
        "\n",
        "\n",
        "mkdir -p trimmed_fastq_files\n",
        "ls -F"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "fastq_files/\n",
            "miniconda.sh\n",
            "sample_data/\n",
            "trimmed_fastq_files/\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9bbe2bd9"
      },
      "source": [
        "Now, we will activate the `workshop_ngs` conda environment and run `Trim Galore!` on all the FASTQ files. We'll specify that they are paired-end reads and direct the output to the `trimmed_fastq_files` directory.\n",
        "\n",
        "Trim Galore automatically detects and removes common adapter sequences and performs quality trimming. The `--paired` option is essential for correctly processing paired-end reads."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "import os\n",
        "\n",
        "# Direct paths for this single sample\n",
        "read1 = \"fastq_files/BEN_CI16_sub_1.fq.gz\"\n",
        "read2 = \"fastq_files/BEN_CI16_sub_2.fq.gz\"\n",
        "output_dir = \"trimmed_fastq_files\"\n",
        "\n",
        "# Make output directory if missing\n",
        "os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "print(\"\\n--- Running Trim Galore for BEN_CI16 ---\")\n",
        "\n",
        "# Run Trim Galore inside conda environment\n",
        "!bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        "          conda activate workshop_ngs && \\\n",
        "          trim_galore --paired {read1} {read2} -o {output_dir}\"\n",
        "\n",
        "# Remove unwanted reports and QC files\n",
        "!rm -f {output_dir}/BEN_CI16_sub_*.fq.gz_trimming_report.txt\n",
        "!rm -f {output_dir}/BEN_CI16_sub_*_fastqc.html\n",
        "!rm -f {output_dir}/BEN_CI16_sub_*_fastqc.zip\n",
        "\n",
        "# Show trimmed output files\n",
        "print('\\n--- Trimmed files in trimmed_fastq_files/ ---')\n",
        "!ls -F trimmed_fastq_files/\n"
      ],
      "metadata": {
        "id": "uBB4jsmR6Xu_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "# Input FASTQ files\n",
        "read1=\"fastq_files/BEN_CI16_sub_1.fq.gz\"\n",
        "read2=\"fastq_files/BEN_CI16_sub_2.fq.gz\"\n",
        "output_dir=\"trimmed_fastq_files\"\n",
        "\n",
        "echo -e \"\\n--- Running Trim Galore for BEN_CI16 ---\"\n",
        "\n",
        "# Load conda and activate environment\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate workshop_ngs\n",
        "\n",
        "# Run Trim Galore (directory already exists)\n",
        "trim_galore --paired \"${read1}\" \"${read2}\" -o \"${output_dir}\"\n",
        "\n",
        "# Remove unnecessary reports\n",
        "rm -f ${output_dir}/BEN_CI16_sub_*.fq.gz_trimming_report.txt\n",
        "rm -f ${output_dir}/BEN_CI16_sub_*_fastqc.html\n",
        "rm -f ${output_dir}/BEN_CI16_sub_*_fastqc.zip\n",
        "\n",
        "echo -e \"\\n--- Trimmed files in trimmed_fastq_files/ ---\"\n",
        "ls -F trimmed_fastq_files/\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "wXFX7QfcTbM3",
        "outputId": "3043c997-fb8e-417c-9dde-3cc61e641d05"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "--- Running Trim Galore for BEN_CI16 ---\n",
            "igzip command line interface 2.31.1\n",
            "\n",
            "--- Trimmed files in trimmed_fastq_files/ ---\n",
            "BEN_CI16_sub_1_val_1.fq.gz\n",
            "BEN_CI16_sub_2_val_2.fq.gz\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Multicore support not enabled. Proceeding with single-core trimming.\n",
            "Path to Cutadapt set as: 'cutadapt' (default)\n",
            "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
            "Cutadapt version: 4.4\n",
            "single-core operation.\n",
            "igzip detected. Using igzip for decompressing\n",
            "\n",
            "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
            "\n",
            "Output will be written into the directory: /content/trimmed_fastq_files/\n",
            "\n",
            "\n",
            "AUTO-DETECTING ADAPTER TYPE\n",
            "===========================\n",
            "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> fastq_files/BEN_CI16_sub_1.fq.gz <<)\n",
            "\n",
            "Found perfect matches for the following adapter sequences:\n",
            "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
            "Illumina\t176422\tAGATCGGAAGAGC\t1000000\t17.64\n",
            "Nextera\t11\tCTGTCTCTTATA\t1000000\t0.00\n",
            "smallRNA\t6\tTGGAATTCTCGG\t1000000\t0.00\n",
            "Using Illumina adapter for trimming (count: 176422). Second best hit was Nextera (count: 11)\n",
            "\n",
            "Writing report to '/content/trimmed_fastq_files/BEN_CI16_sub_1.fq.gz_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: fastq_files/BEN_CI16_sub_1.fq.gz\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.4\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "Output file(s) will be GZIP compressed\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.4). Setting -j 1\n",
            "Writing final adapter and quality trimmed output to BEN_CI16_sub_1_trimmed.fq.gz\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file fastq_files/BEN_CI16_sub_1.fq.gz <<< \n",
            "This is cutadapt 4.4 with Python 3.7.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC fastq_files/BEN_CI16_sub_1.fq.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 62.578 s (62.578 s/read; 0.96 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:               1,000,000\n",
            "Reads with adapters:                   504,019 (50.4%)\n",
            "Reads written (passing filters):     1,000,000 (100.0%)\n",
            "\n",
            "Total basepairs processed:   150,000,000 bp\n",
            "Quality-trimmed:               1,215,638 bp (0.8%)\n",
            "Total written (filtered):    141,006,039 bp (94.0%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 504019 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 28.6%\n",
            "  C: 26.0%\n",
            "  G: 19.6%\n",
            "  T: 25.7%\n",
            "  none/other: 0.1%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t197827\t250000.0\t0\t197827\n",
            "2\t57274\t62500.0\t0\t57274\n",
            "3\t21011\t15625.0\t0\t21011\n",
            "4\t7894\t3906.2\t0\t7894\n",
            "5\t5348\t976.6\t0\t5348\n",
            "6\t4796\t244.1\t0\t4796\n",
            "7\t4581\t61.0\t0\t4581\n",
            "8\t4266\t15.3\t0\t4266\n",
            "9\t4233\t3.8\t0\t4212 21\n",
            "10\t4660\t1.0\t1\t4438 222\n",
            "11\t4394\t0.2\t1\t4198 196\n",
            "12\t4506\t0.1\t1\t4292 214\n",
            "13\t4419\t0.0\t1\t4192 227\n",
            "14\t4250\t0.0\t1\t4022 228\n",
            "15\t4234\t0.0\t1\t4011 223\n",
            "16\t4227\t0.0\t1\t4006 221\n",
            "17\t4460\t0.0\t1\t4163 297\n",
            "18\t4404\t0.0\t1\t4153 251\n",
            "19\t4101\t0.0\t1\t3934 167\n",
            "20\t4227\t0.0\t1\t4013 214\n",
            "21\t4038\t0.0\t1\t3826 212\n",
            "22\t4066\t0.0\t1\t3867 199\n",
            "23\t3935\t0.0\t1\t3753 182\n",
            "24\t3928\t0.0\t1\t3749 179\n",
            "25\t3896\t0.0\t1\t3694 202\n",
            "26\t3801\t0.0\t1\t3641 160\n",
            "27\t3898\t0.0\t1\t3723 175\n",
            "28\t4056\t0.0\t1\t3834 222\n",
            "29\t3832\t0.0\t1\t3659 173\n",
            "30\t3809\t0.0\t1\t3627 182\n",
            "31\t3547\t0.0\t1\t3402 145\n",
            "32\t3515\t0.0\t1\t3372 143\n",
            "33\t3249\t0.0\t1\t3145 104\n",
            "34\t3491\t0.0\t1\t3341 150\n",
            "35\t3362\t0.0\t1\t3219 143\n",
            "36\t3360\t0.0\t1\t3215 145\n",
            "37\t3276\t0.0\t1\t3149 127\n",
            "38\t3283\t0.0\t1\t3155 128\n",
            "39\t3376\t0.0\t1\t3268 108\n",
            "40\t3469\t0.0\t1\t3327 142\n",
            "41\t3454\t0.0\t1\t3329 125\n",
            "42\t2902\t0.0\t1\t2801 101\n",
            "43\t4048\t0.0\t1\t3925 123\n",
            "44\t1747\t0.0\t1\t1693 54\n",
            "45\t2432\t0.0\t1\t2357 75\n",
            "46\t2650\t0.0\t1\t2570 80\n",
            "47\t2628\t0.0\t1\t2554 74\n",
            "48\t2528\t0.0\t1\t2456 72\n",
            "49\t2424\t0.0\t1\t2355 69\n",
            "50\t2487\t0.0\t1\t2416 71\n",
            "51\t2497\t0.0\t1\t2410 87\n",
            "52\t2330\t0.0\t1\t2254 76\n",
            "53\t2498\t0.0\t1\t2440 58\n",
            "54\t1986\t0.0\t1\t1940 46\n",
            "55\t1958\t0.0\t1\t1892 66\n",
            "56\t1850\t0.0\t1\t1791 59\n",
            "57\t1783\t0.0\t1\t1737 46\n",
            "58\t1912\t0.0\t1\t1868 44\n",
            "59\t1870\t0.0\t1\t1808 62\n",
            "60\t1757\t0.0\t1\t1711 46\n",
            "61\t1851\t0.0\t1\t1804 47\n",
            "62\t1791\t0.0\t1\t1741 50\n",
            "63\t1433\t0.0\t1\t1398 35\n",
            "64\t1756\t0.0\t1\t1704 52\n",
            "65\t1382\t0.0\t1\t1353 29\n",
            "66\t1356\t0.0\t1\t1324 32\n",
            "67\t1485\t0.0\t1\t1444 41\n",
            "68\t1591\t0.0\t1\t1560 31\n",
            "69\t2210\t0.0\t1\t1964 246\n",
            "70\t8264\t0.0\t1\t8107 157\n",
            "71\t1531\t0.0\t1\t1478 53\n",
            "72\t375\t0.0\t1\t351 24\n",
            "73\t259\t0.0\t1\t243 16\n",
            "74\t184\t0.0\t1\t174 10\n",
            "75\t171\t0.0\t1\t163 8\n",
            "76\t156\t0.0\t1\t146 10\n",
            "77\t146\t0.0\t1\t143 3\n",
            "78\t176\t0.0\t1\t170 6\n",
            "79\t184\t0.0\t1\t173 11\n",
            "80\t198\t0.0\t1\t190 8\n",
            "81\t167\t0.0\t1\t161 6\n",
            "82\t146\t0.0\t1\t144 2\n",
            "83\t96\t0.0\t1\t93 3\n",
            "84\t78\t0.0\t1\t72 6\n",
            "85\t75\t0.0\t1\t72 3\n",
            "86\t73\t0.0\t1\t62 11\n",
            "87\t64\t0.0\t1\t63 1\n",
            "88\t57\t0.0\t1\t53 4\n",
            "89\t51\t0.0\t1\t49 2\n",
            "90\t64\t0.0\t1\t61 3\n",
            "91\t62\t0.0\t1\t59 3\n",
            "92\t46\t0.0\t1\t46\n",
            "93\t36\t0.0\t1\t36\n",
            "94\t25\t0.0\t1\t22 3\n",
            "95\t30\t0.0\t1\t27 3\n",
            "96\t27\t0.0\t1\t25 2\n",
            "97\t31\t0.0\t1\t30 1\n",
            "98\t18\t0.0\t1\t18\n",
            "99\t14\t0.0\t1\t12 2\n",
            "100\t26\t0.0\t1\t25 1\n",
            "101\t22\t0.0\t1\t21 1\n",
            "102\t17\t0.0\t1\t16 1\n",
            "103\t13\t0.0\t1\t12 1\n",
            "104\t14\t0.0\t1\t11 3\n",
            "105\t6\t0.0\t1\t4 2\n",
            "106\t14\t0.0\t1\t13 1\n",
            "107\t14\t0.0\t1\t13 1\n",
            "108\t6\t0.0\t1\t6\n",
            "109\t9\t0.0\t1\t6 3\n",
            "110\t7\t0.0\t1\t4 3\n",
            "111\t14\t0.0\t1\t10 4\n",
            "112\t4\t0.0\t1\t3 1\n",
            "113\t2\t0.0\t1\t0 2\n",
            "114\t2\t0.0\t1\t2\n",
            "115\t2\t0.0\t1\t2\n",
            "116\t5\t0.0\t1\t3 2\n",
            "117\t1\t0.0\t1\t0 1\n",
            "118\t4\t0.0\t1\t2 2\n",
            "119\t2\t0.0\t1\t2\n",
            "120\t3\t0.0\t1\t3\n",
            "122\t1\t0.0\t1\t0 1\n",
            "123\t1\t0.0\t1\t0 1\n",
            "124\t1\t0.0\t1\t1\n",
            "125\t1\t0.0\t1\t0 1\n",
            "126\t1\t0.0\t1\t1\n",
            "127\t2\t0.0\t1\t2\n",
            "128\t4\t0.0\t1\t2 2\n",
            "129\t1\t0.0\t1\t0 1\n",
            "130\t4\t0.0\t1\t2 2\n",
            "131\t3\t0.0\t1\t3\n",
            "132\t4\t0.0\t1\t4\n",
            "133\t1\t0.0\t1\t0 1\n",
            "134\t5\t0.0\t1\t5\n",
            "135\t3\t0.0\t1\t0 3\n",
            "136\t3\t0.0\t1\t3\n",
            "139\t1\t0.0\t1\t0 1\n",
            "140\t1\t0.0\t1\t0 1\n",
            "141\t1\t0.0\t1\t0 1\n",
            "144\t1\t0.0\t1\t0 1\n",
            "145\t1\t0.0\t1\t0 1\n",
            "146\t1\t0.0\t1\t0 1\n",
            "147\t3\t0.0\t1\t0 3\n",
            "148\t4\t0.0\t1\t0 4\n",
            "149\t13\t0.0\t1\t0 13\n",
            "150\t72\t0.0\t1\t0 72\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: fastq_files/BEN_CI16_sub_1.fq.gz\n",
            "=============================================\n",
            "1000000 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Writing report to '/content/trimmed_fastq_files/BEN_CI16_sub_2.fq.gz_trimming_report.txt'\n",
            "\n",
            "SUMMARISING RUN PARAMETERS\n",
            "==========================\n",
            "Input filename: fastq_files/BEN_CI16_sub_2.fq.gz\n",
            "Trimming mode: paired-end\n",
            "Trim Galore version: 0.6.10\n",
            "Cutadapt version: 4.4\n",
            "Number of cores used for trimming: 1\n",
            "Quality Phred score cutoff: 20\n",
            "Quality encoding type selected: ASCII+33\n",
            "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
            "Maximum trimming error rate: 0.1 (default)\n",
            "Minimum required adapter overlap (stringency): 1 bp\n",
            "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
            "Output file(s) will be GZIP compressed\n",
            "\n",
            "Cutadapt seems to be fairly up-to-date (version 4.4). Setting -j -j 1\n",
            "Writing final adapter and quality trimmed output to BEN_CI16_sub_2_trimmed.fq.gz\n",
            "\n",
            "\n",
            "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file fastq_files/BEN_CI16_sub_2.fq.gz <<< \n",
            "This is cutadapt 4.4 with Python 3.7.12\n",
            "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC fastq_files/BEN_CI16_sub_2.fq.gz\n",
            "Processing single-end reads on 1 core ...\n",
            "Finished in 68.294 s (68.294 s/read; 0.88 M reads/minute).\n",
            "\n",
            "=== Summary ===\n",
            "\n",
            "Total reads processed:               1,000,000\n",
            "Reads with adapters:                   476,456 (47.6%)\n",
            "Reads written (passing filters):     1,000,000 (100.0%)\n",
            "\n",
            "Total basepairs processed:   150,000,000 bp\n",
            "Quality-trimmed:               6,784,111 bp (4.5%)\n",
            "Total written (filtered):    136,180,090 bp (90.8%)\n",
            "\n",
            "=== Adapter 1 ===\n",
            "\n",
            "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 476456 times\n",
            "\n",
            "Minimum overlap: 1\n",
            "No. of allowed errors:\n",
            "1-9 bp: 0; 10-13 bp: 1\n",
            "\n",
            "Bases preceding removed adapters:\n",
            "  A: 28.2%\n",
            "  C: 26.9%\n",
            "  G: 19.1%\n",
            "  T: 25.6%\n",
            "  none/other: 0.1%\n",
            "\n",
            "Overview of removed sequences\n",
            "length\tcount\texpect\tmax.err\terror counts\n",
            "1\t190832\t250000.0\t0\t190832\n",
            "2\t56994\t62500.0\t0\t56994\n",
            "3\t19626\t15625.0\t0\t19626\n",
            "4\t6950\t3906.2\t0\t6950\n",
            "5\t5247\t976.6\t0\t5247\n",
            "6\t4651\t244.1\t0\t4651\n",
            "7\t4334\t61.0\t0\t4334\n",
            "8\t3786\t15.3\t0\t3786\n",
            "9\t3475\t3.8\t0\t3456 19\n",
            "10\t4268\t1.0\t1\t3824 444\n",
            "11\t3860\t0.2\t1\t3457 403\n",
            "12\t4243\t0.1\t1\t3696 547\n",
            "13\t4433\t0.0\t1\t3864 569\n",
            "14\t6360\t0.0\t1\t5516 844\n",
            "15\t2790\t0.0\t1\t2513 277\n",
            "16\t4156\t0.0\t1\t3714 442\n",
            "17\t4219\t0.0\t1\t3794 425\n",
            "18\t2619\t0.0\t1\t2359 260\n",
            "19\t4135\t0.0\t1\t3727 408\n",
            "20\t3172\t0.0\t1\t2883 289\n",
            "21\t2997\t0.0\t1\t2726 271\n",
            "22\t3683\t0.0\t1\t3350 333\n",
            "23\t3792\t0.0\t1\t3422 370\n",
            "24\t5040\t0.0\t1\t4513 527\n",
            "25\t3118\t0.0\t1\t2871 247\n",
            "26\t3075\t0.0\t1\t2834 241\n",
            "27\t2855\t0.0\t1\t2644 211\n",
            "28\t3713\t0.0\t1\t3370 343\n",
            "29\t3128\t0.0\t1\t2854 274\n",
            "30\t4811\t0.0\t1\t4414 397\n",
            "31\t2255\t0.0\t1\t2096 159\n",
            "32\t3474\t0.0\t1\t3192 282\n",
            "33\t2886\t0.0\t1\t2648 238\n",
            "34\t2823\t0.0\t1\t2611 212\n",
            "35\t3199\t0.0\t1\t2936 263\n",
            "36\t3033\t0.0\t1\t2822 211\n",
            "37\t2819\t0.0\t1\t2625 194\n",
            "38\t3265\t0.0\t1\t3025 240\n",
            "39\t2955\t0.0\t1\t2755 200\n",
            "40\t3250\t0.0\t1\t2956 294\n",
            "41\t3369\t0.0\t1\t3100 269\n",
            "42\t4150\t0.0\t1\t3861 289\n",
            "43\t2216\t0.0\t1\t2082 134\n",
            "44\t3244\t0.0\t1\t3020 224\n",
            "45\t3147\t0.0\t1\t2931 216\n",
            "46\t1981\t0.0\t1\t1858 123\n",
            "47\t2162\t0.0\t1\t2040 122\n",
            "48\t2813\t0.0\t1\t2666 147\n",
            "49\t2642\t0.0\t1\t2544 98\n",
            "50\t2120\t0.0\t1\t1997 123\n",
            "51\t4061\t0.0\t1\t3883 178\n",
            "52\t1245\t0.0\t1\t1189 56\n",
            "53\t1624\t0.0\t1\t1569 55\n",
            "54\t1193\t0.0\t1\t1122 71\n",
            "55\t2008\t0.0\t1\t1915 93\n",
            "56\t1999\t0.0\t1\t1927 72\n",
            "57\t1714\t0.0\t1\t1641 73\n",
            "58\t1739\t0.0\t1\t1670 69\n",
            "59\t1774\t0.0\t1\t1696 78\n",
            "60\t1697\t0.0\t1\t1640 57\n",
            "61\t1931\t0.0\t1\t1866 65\n",
            "62\t2282\t0.0\t1\t2196 86\n",
            "63\t3072\t0.0\t1\t2782 290\n",
            "64\t8648\t0.0\t1\t8410 238\n",
            "65\t1787\t0.0\t1\t1715 72\n",
            "66\t570\t0.0\t1\t526 44\n",
            "67\t443\t0.0\t1\t405 38\n",
            "68\t316\t0.0\t1\t299 17\n",
            "69\t315\t0.0\t1\t299 16\n",
            "70\t354\t0.0\t1\t335 19\n",
            "71\t285\t0.0\t1\t263 22\n",
            "72\t300\t0.0\t1\t277 23\n",
            "73\t260\t0.0\t1\t240 20\n",
            "74\t251\t0.0\t1\t233 18\n",
            "75\t229\t0.0\t1\t213 16\n",
            "76\t194\t0.0\t1\t183 11\n",
            "77\t164\t0.0\t1\t153 11\n",
            "78\t135\t0.0\t1\t127 8\n",
            "79\t141\t0.0\t1\t129 12\n",
            "80\t158\t0.0\t1\t154 4\n",
            "81\t149\t0.0\t1\t137 12\n",
            "82\t132\t0.0\t1\t121 11\n",
            "83\t98\t0.0\t1\t92 6\n",
            "84\t81\t0.0\t1\t75 6\n",
            "85\t70\t0.0\t1\t67 3\n",
            "86\t67\t0.0\t1\t62 5\n",
            "87\t68\t0.0\t1\t61 7\n",
            "88\t57\t0.0\t1\t54 3\n",
            "89\t55\t0.0\t1\t49 6\n",
            "90\t68\t0.0\t1\t60 8\n",
            "91\t42\t0.0\t1\t38 4\n",
            "92\t42\t0.0\t1\t37 5\n",
            "93\t29\t0.0\t1\t25 4\n",
            "94\t30\t0.0\t1\t26 4\n",
            "95\t22\t0.0\t1\t19 3\n",
            "96\t19\t0.0\t1\t17 2\n",
            "97\t23\t0.0\t1\t22 1\n",
            "98\t20\t0.0\t1\t18 2\n",
            "99\t18\t0.0\t1\t17 1\n",
            "100\t26\t0.0\t1\t25 1\n",
            "101\t18\t0.0\t1\t18\n",
            "102\t15\t0.0\t1\t12 3\n",
            "103\t16\t0.0\t1\t13 3\n",
            "104\t20\t0.0\t1\t18 2\n",
            "105\t13\t0.0\t1\t12 1\n",
            "106\t9\t0.0\t1\t7 2\n",
            "107\t10\t0.0\t1\t9 1\n",
            "108\t5\t0.0\t1\t5\n",
            "109\t6\t0.0\t1\t5 1\n",
            "110\t6\t0.0\t1\t5 1\n",
            "111\t12\t0.0\t1\t9 3\n",
            "112\t6\t0.0\t1\t4 2\n",
            "113\t8\t0.0\t1\t6 2\n",
            "114\t3\t0.0\t1\t2 1\n",
            "115\t5\t0.0\t1\t4 1\n",
            "116\t5\t0.0\t1\t5\n",
            "118\t5\t0.0\t1\t2 3\n",
            "119\t1\t0.0\t1\t1\n",
            "120\t2\t0.0\t1\t1 1\n",
            "122\t1\t0.0\t1\t0 1\n",
            "123\t2\t0.0\t1\t1 1\n",
            "124\t1\t0.0\t1\t0 1\n",
            "126\t2\t0.0\t1\t2\n",
            "127\t1\t0.0\t1\t1\n",
            "128\t3\t0.0\t1\t2 1\n",
            "130\t3\t0.0\t1\t2 1\n",
            "131\t4\t0.0\t1\t2 2\n",
            "132\t11\t0.0\t1\t11\n",
            "134\t11\t0.0\t1\t9 2\n",
            "136\t5\t0.0\t1\t4 1\n",
            "137\t2\t0.0\t1\t0 2\n",
            "139\t1\t0.0\t1\t0 1\n",
            "140\t2\t0.0\t1\t1 1\n",
            "141\t2\t0.0\t1\t0 2\n",
            "142\t2\t0.0\t1\t0 2\n",
            "145\t1\t0.0\t1\t1\n",
            "147\t1\t0.0\t1\t1\n",
            "148\t2\t0.0\t1\t0 2\n",
            "149\t3\t0.0\t1\t0 3\n",
            "150\t61\t0.0\t1\t0 61\n",
            "\n",
            "RUN STATISTICS FOR INPUT FILE: fastq_files/BEN_CI16_sub_2.fq.gz\n",
            "=============================================\n",
            "1000000 sequences processed in total\n",
            "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
            "\n",
            "Validate paired-end files BEN_CI16_sub_1_trimmed.fq.gz and BEN_CI16_sub_2_trimmed.fq.gz\n",
            "file_1: BEN_CI16_sub_1_trimmed.fq.gz, file_2: BEN_CI16_sub_2_trimmed.fq.gz\n",
            "\n",
            "\n",
            ">>>>> Now validing the length of the 2 paired-end infiles: BEN_CI16_sub_1_trimmed.fq.gz and BEN_CI16_sub_2_trimmed.fq.gz <<<<<\n",
            "Writing validated paired-end Read 1 reads to BEN_CI16_sub_1_val_1.fq.gz\n",
            "Writing validated paired-end Read 2 reads to BEN_CI16_sub_2_val_2.fq.gz\n",
            "\n",
            "Total number of sequences analysed: 1000000\n",
            "\n",
            "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 5281 (0.53%)\n",
            "\n",
            "Deleting both intermediate output files BEN_CI16_sub_1_trimmed.fq.gz and BEN_CI16_sub_2_trimmed.fq.gz\n",
            "\n",
            "====================================================================================================\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9ead2635"
      },
      "source": [
        "### Downloading reference files for mapping\n",
        "\n",
        "To prepare for the read mapping step, we need to download the reference genome and its associated index files. These files will be stored in a new directory called `reference`."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "!rm -rf reference\n",
        "reference_tar = \"https://zenodo.org/records/17878528/files/reference.tar.gz\"\n",
        "\n",
        "# Download the tar.gz file\n",
        "!wget -q $reference_tar -O reference.tar.gz\n",
        "\n",
        "# Extract it  this will unpack into its own folder (likely 'reference/')\n",
        "!tar -xzvf reference.tar.gz\n",
        "\n",
        "# List extracted contents\n",
        "!ls -F reference/"
      ],
      "metadata": {
        "id": "0qiw_azslQFG",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "43211aff-a6a9-447d-d046-e6f908d15705"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "reference/\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.ann\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.amb\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.sa\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.bwt\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.pac\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
            "reference/.ipynb_checkpoints/\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.amb\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.ann\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.bwt\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.pac\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.sa\n",
            "CPU times: user 1.83 s, sys: 320 ms, total: 2.15 s\n",
            "Wall time: 10min 25s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "# Remove any existing reference directory\n",
        "rm -rf reference\n",
        "\n",
        "# URL of the reference tarball\n",
        "reference_tar=\"https://zenodo.org/records/17878528/files/reference.tar.gz\"\n",
        "\n",
        "# Download the tar.gz file\n",
        "wget -q \"${reference_tar}\" -O reference.tar.gz\n",
        "\n",
        "# Extract it  it will unpack into \"reference/\"\n",
        "tar -xzvf reference.tar.gz\n",
        "\n",
        "# List extracted contents\n",
        "ls -F reference/\n"
      ],
      "metadata": {
        "id": "3zeKpXwqZlvz",
        "outputId": "bead263d-4dc0-4c72-8dc3-60395d56fbe3",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "reference/\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.ann\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.amb\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.sa\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.bwt\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna.pac\n",
            "reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
            "reference/.ipynb_checkpoints/\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.amb\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.ann\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.bwt\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.pac\n",
            "GCA_021130815.1_PanTigT.MC.v3_genomic.fna.sa\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "ls -ltrh reference/"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "u2NDUH7JAlIR",
        "outputId": "be7b533c-03f8-4c1c-c3ca-3e3c039ed058"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total 6.2G\n",
            "-rw-r--r-- 1 root root 2.3G Dec  9 11:52 GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
            "-rw-r--r-- 1 root root 2.3G Dec  9 13:02 GCA_021130815.1_PanTigT.MC.v3_genomic.fna.bwt\n",
            "-rw-r--r-- 1 root root 574M Dec  9 13:02 GCA_021130815.1_PanTigT.MC.v3_genomic.fna.pac\n",
            "-rw-r--r-- 1 root root 137K Dec  9 13:02 GCA_021130815.1_PanTigT.MC.v3_genomic.fna.ann\n",
            "-rw-r--r-- 1 root root  20K Dec  9 13:02 GCA_021130815.1_PanTigT.MC.v3_genomic.fna.amb\n",
            "-rw-r--r-- 1 root root 1.2G Dec  9 13:35 GCA_021130815.1_PanTigT.MC.v3_genomic.fna.sa\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ea38b6a9"
      },
      "source": [
        "# Mapping reads to the Reference genome:\n",
        "Here we:\n",
        "\n",
        "Use an aligner like BWA or Bowtie2 to map the cleaned reads to a reference genome\n",
        "\n",
        "Convert output from SAM to BAM format (compressed and binary)\n",
        "\n",
        "Sort the BAM files by genomic coordinates (using samtools sort)\n",
        "\n",
        "\n",
        "Index the BAM files so tools can access them efficiently\n",
        "\n",
        "*Why it's important:* Proper alignment is the foundation for all downstream analysis. Sorting/indexing ensures quick and efficient variant detection.\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Create a new directory named `mapped_reads` to store output alignment files, then index the reference genome located at `reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna` using `bwa index` for efficient read mapping. After indexing, map the trimmed paired-end reads from each sample to the indexed reference genome using `bwa mem`, converting the output to sorted BAM files and indexing them using `samtools`."
      ],
      "metadata": {
        "id": "rlOQNcpAF1ws"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir -p mapped_reads\n",
        "\n",
        "# Paths for this sample\n",
        "REFERENCE_GENOME = \"reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna\"\n",
        "\n",
        "read1 = \"trimmed_fastq_files/BEN_CI16_sub_1_val_1.fq.gz\"\n",
        "read2 = \"trimmed_fastq_files/BEN_CI16_sub_2_val_2.fq.gz\"\n",
        "output_bam = \"mapped_reads/BEN_CI16.bam\"\n",
        "\n",
        "print(\"\\n--- Mapping BEN_CI16 ---\")\n",
        "\n",
        "# Map + convert to BAM + sort BAM\n",
        "!bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        "          conda activate workshop_ngs && \\\n",
        "          bwa mem -M -R '@RG\\tID:BEN_CI16\\tSM:BEN_CI16\\tPL:ILLUMINA' \\\n",
        "          {REFERENCE_GENOME} {read1} {read2} | \\\n",
        "          samtools view -bS - | \\\n",
        "          samtools sort -o {output_bam} -\"\n",
        "\n",
        "# Index sorted BAM\n",
        "#print(\"Indexing BEN_CI16...\")\n",
        "#!bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        " #         conda activate workshop_ngs && \\\n",
        "  #        samtools index {output_bam}\"\n",
        "\n",
        "# List output files\n",
        "!ls -lh mapped_reads/\n"
      ],
      "metadata": {
        "id": "J56P3nBKnGm8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "# Create output directory (safe even if it already exists)\n",
        "mkdir -p mapped_reads\n",
        "\n",
        "# Paths for this sample\n",
        "REFERENCE_GENOME=\"reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna\"\n",
        "read1=\"trimmed_fastq_files/BEN_CI16_sub_1_val_1.fq.gz\"\n",
        "read2=\"trimmed_fastq_files/BEN_CI16_sub_2_val_2.fq.gz\"\n",
        "output_bam=\"mapped_reads/BEN_CI16.bam\"\n",
        "\n",
        "echo -e \"\\n--- Mapping BEN_CI16 ---\"\n",
        "\n",
        "# Load conda and activate environment\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate workshop_ngs\n",
        "\n",
        "# Map  convert SAMBAM  sort BAM\n",
        "bwa mem -M -R '@RG\\tID:BEN_CI16\\tSM:BEN_CI16\\tPL:ILLUMINA' \\\n",
        "    \"$REFERENCE_GENOME\" \"$read1\" \"$read2\" \\\n",
        "    | samtools view -bS - \\\n",
        "    | samtools sort -o \"$output_bam\" -\n",
        "\n",
        "# DO NOT INDEX YET (only after MarkDuplicates)\n",
        "\n",
        "# List output files\n",
        "ls -lh mapped_reads/\n"
      ],
      "metadata": {
        "id": "9lg9DFTAaYPt",
        "outputId": "fbd6e03e-0f9b-4c47-cb0c-789d0a587558",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "--- Mapping BEN_CI16 ---\n",
            "total 203M\n",
            "-rw-r--r-- 1 root root 203M Dec 12 12:37 BEN_CI16.bam\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 72058 sequences (10000070 bp)...\n",
            "[M::process] read 72038 sequences (10000031 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32094, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.18, 86.53)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 72058 reads in 35.713 CPU sec, 38.300 real sec\n",
            "[M::process] read 71952 sequences (10000278 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 32092, 2, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 262)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 488)\n",
            "[M::mem_pestat] mean and std.dev: (209.58, 84.44)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 601)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 72038 reads in 32.612 CPU sec, 32.748 real sec\n",
            "[M::process] read 71944 sequences (10000256 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32060, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.25, 85.83)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71952 reads in 32.269 CPU sec, 32.373 real sec\n",
            "[M::process] read 72006 sequences (10000134 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32136, 0, 1)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.34, 85.96)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71944 reads in 33.495 CPU sec, 33.854 real sec\n",
            "[M::process] read 71902 sequences (10000027 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32060, 2, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (148, 196, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 499)\n",
            "[M::mem_pestat] mean and std.dev: (211.39, 86.69)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 616)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 72006 reads in 32.962 CPU sec, 33.407 real sec\n",
            "[M::process] read 71936 sequences (10000194 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32089, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.70, 86.31)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71902 reads in 32.533 CPU sec, 32.620 real sec\n",
            "[M::process] read 71860 sequences (10000033 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32154, 0, 1)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 264)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 494)\n",
            "[M::mem_pestat] mean and std.dev: (211.09, 85.51)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 609)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71936 reads in 33.101 CPU sec, 33.182 real sec\n",
            "[M::process] read 71916 sequences (10000145 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32060, 1, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 197, 266)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 500)\n",
            "[M::mem_pestat] mean and std.dev: (212.00, 85.98)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 617)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71860 reads in 32.606 CPU sec, 33.550 real sec\n",
            "[M::process] read 71882 sequences (10000280 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (2, 32150, 2, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 263)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 491)\n",
            "[M::mem_pestat] mean and std.dev: (210.36, 84.71)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 605)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71916 reads in 32.585 CPU sec, 32.629 real sec\n",
            "[M::process] read 72004 sequences (10000085 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 31997, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (150, 197, 264)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 492)\n",
            "[M::mem_pestat] mean and std.dev: (211.46, 85.24)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 606)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71882 reads in 33.097 CPU sec, 33.184 real sec\n",
            "[M::process] read 71932 sequences (10000107 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 32178, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.16, 85.90)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 72004 reads in 33.486 CPU sec, 33.609 real sec\n",
            "[M::process] read 71908 sequences (10000194 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32176, 2, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 197, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.48, 85.98)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71932 reads in 32.770 CPU sec, 33.617 real sec\n",
            "[M::process] read 71946 sequences (10000137 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32075, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 197, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.40, 85.86)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71908 reads in 32.640 CPU sec, 32.666 real sec\n",
            "[M::process] read 71980 sequences (10000263 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32154, 1, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 263)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 491)\n",
            "[M::mem_pestat] mean and std.dev: (210.12, 85.58)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 605)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71946 reads in 33.315 CPU sec, 33.503 real sec\n",
            "[M::process] read 71958 sequences (10000204 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 32057, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.33, 86.05)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71980 reads in 33.856 CPU sec, 34.866 real sec\n",
            "[M::process] read 71974 sequences (10000032 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 32078, 0, 1)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (150, 197, 267)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 501)\n",
            "[M::mem_pestat] mean and std.dev: (211.94, 86.30)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 618)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71958 reads in 33.507 CPU sec, 33.639 real sec\n",
            "[M::process] read 71988 sequences (10000233 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32085, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.05, 86.07)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71974 reads in 33.333 CPU sec, 33.381 real sec\n",
            "[M::process] read 71914 sequences (10000156 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 31902, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 197, 266)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 500)\n",
            "[M::mem_pestat] mean and std.dev: (211.92, 86.29)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 617)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71988 reads in 38.590 CPU sec, 39.817 real sec\n",
            "[M::process] read 71936 sequences (10000039 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32007, 0, 1)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.88, 86.83)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71914 reads in 33.550 CPU sec, 34.410 real sec\n",
            "[M::process] read 72028 sequences (10000265 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 32108, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 195, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (210.96, 86.34)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71936 reads in 33.403 CPU sec, 34.473 real sec\n",
            "[M::process] read 71978 sequences (10000037 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32123, 0, 2)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 264)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 494)\n",
            "[M::mem_pestat] mean and std.dev: (210.93, 86.05)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 609)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 72028 reads in 33.428 CPU sec, 33.511 real sec\n",
            "[M::process] read 71942 sequences (10000183 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32107, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (148, 195, 264)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 496)\n",
            "[M::mem_pestat] mean and std.dev: (210.94, 86.11)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 612)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71978 reads in 32.808 CPU sec, 32.944 real sec\n",
            "[M::process] read 71938 sequences (10000259 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (3, 32025, 0, 1)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 197, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.73, 86.26)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71942 reads in 33.650 CPU sec, 34.780 real sec\n",
            "[M::process] read 71968 sequences (10000222 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32125, 1, 1)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 197, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.70, 85.79)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71938 reads in 32.922 CPU sec, 33.004 real sec\n",
            "[M::process] read 72020 sequences (10000256 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 32061, 1, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 197, 266)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 500)\n",
            "[M::mem_pestat] mean and std.dev: (212.28, 87.04)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 617)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71968 reads in 33.016 CPU sec, 33.102 real sec\n",
            "[M::process] read 71868 sequences (10000113 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 32094, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (148, 196, 264)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 496)\n",
            "[M::mem_pestat] mean and std.dev: (211.03, 86.01)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 612)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 72020 reads in 34.582 CPU sec, 34.937 real sec\n",
            "[M::process] read 46662 sequences (6489269 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 32197, 4, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 197, 265)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 497)\n",
            "[M::mem_pestat] mean and std.dev: (211.41, 85.40)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 613)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 71868 reads in 31.990 CPU sec, 32.108 real sec\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 20858, 0, 0)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (149, 196, 264)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 494)\n",
            "[M::mem_pestat] mean and std.dev: (211.14, 85.49)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 609)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 46662 reads in 20.590 CPU sec, 20.783 real sec\n",
            "[main] Version: 0.7.19-r1273\n",
            "[main] CMD: bwa mem -M -R @RG\\tID:BEN_CI16\\tSM:BEN_CI16\\tPL:ILLUMINA reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna trimmed_fastq_files/BEN_CI16_sub_1_val_1.fq.gz trimmed_fastq_files/BEN_CI16_sub_2_val_2.fq.gz\n",
            "[main] Real time: 959.001 sec; CPU: 928.179 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Mark Duplicates\n",
        "\n",
        "This command detects PCR duplicates and optical duplicates in the aligned reads and removes them. Duplicate reads can bias downstream analyses such as variant calling, so removing them improves data quality.\n",
        "\n",
        "Explanation of the flags we use are given below\n",
        "\n",
        "-I : Input BAM file. This is the sorted BAM produced from the mapping step.\n",
        "\n",
        "-O :\tOutput BAM file after duplicates have been removed.\n",
        "\n",
        "-M :\tFile where GATK writes duplication statistics (number of duplicates, percent duplication, etc.)\n",
        "\n",
        "--REMOVE_DUPLICATES true :\tInstead of just marking duplicates, this flag instructs GATK to remove them entirely from the BAM file.\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "**After mark duplicates we also need to Index the deduplicated bam file**\n",
        "\n",
        "Creates an index (.bai) file for the deduplicated BAM file.\n",
        "This index enables efficient random access to the BAM file and is required for many downstream tools\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "y-jVE9177vj3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        "          conda activate workshop_ngs && \\\n",
        "          gatk MarkDuplicates \\\n",
        "            -I BEN_CI16.bam \\\n",
        "            -O BEN_CI16.deduplicated.bam \\\n",
        "            -M BEN_CI16.metrics.txt \\\n",
        "            --REMOVE_DUPLICATES true\"\n",
        "\n",
        "\n",
        "print(\"\\n--- Indexing deduplicated BAM ---\")\n",
        "\n",
        "!bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        "          conda activate workshop_ngs && \\\n",
        "          samtools index mapped_reads/BEN_CI16.deduplicated.bam\"\n",
        "\n",
        "# List output files\n",
        "!ls -lh mapped_reads/"
      ],
      "metadata": {
        "id": "MtmP9ayY7x-W"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "# Load conda\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate workshop_ngs\n",
        "\n",
        "# Input and output paths\n",
        "input_bam=\"mapped_reads/BEN_CI16.bam\"\n",
        "dedup_bam=\"mapped_reads/BEN_CI16.deduplicated.bam\"\n",
        "metrics_file=\"mapped_reads/BEN_CI16.metrics.txt\"\n",
        "\n",
        "echo -e \"\\n--- Running GATK MarkDuplicates ---\"\n",
        "\n",
        "gatk MarkDuplicates \\\n",
        "    -I \"$input_bam\" \\\n",
        "    -O \"$dedup_bam\" \\\n",
        "    -M \"$metrics_file\" \\\n",
        "    --REMOVE_DUPLICATES true\n",
        "\n",
        "echo -e \"\\n--- Indexing deduplicated BAM ---\"\n",
        "\n",
        "samtools index \"$dedup_bam\"\n",
        "\n",
        "echo -e \"\\n--- Files in mapped_reads/ ---\"\n",
        "ls -lh mapped_reads/\n"
      ],
      "metadata": {
        "id": "DpJ2ashPa31J",
        "outputId": "92f3c01d-0870-42d3-b221-fcd0ea2b9cc8",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "--- Running GATK MarkDuplicates ---\n",
            "Tool returned:\n",
            "0\n",
            "\n",
            "--- Indexing deduplicated BAM ---\n",
            "\n",
            "--- Files in mapped_reads/ ---\n",
            "total 478M\n",
            "-rw-r--r-- 1 root root 203M Dec 12 12:37 BEN_CI16.bam\n",
            "-rw-r--r-- 1 root root 273M Dec 12 12:40 BEN_CI16.deduplicated.bam\n",
            "-rw-r--r-- 1 root root 3.3M Dec 12 12:40 BEN_CI16.deduplicated.bam.bai\n",
            "-rw-r--r-- 1 root root 3.7K Dec 12 12:40 BEN_CI16.metrics.txt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "12:39:48.027 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/usr/local/miniconda/envs/workshop_ngs/share/gatk4-4.3.0.0-0/gatk-package-4.3.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
            "[Fri Dec 12 12:39:48 UTC 2025] MarkDuplicates --INPUT mapped_reads/BEN_CI16.bam --OUTPUT mapped_reads/BEN_CI16.deduplicated.bam --METRICS_FILE mapped_reads/BEN_CI16.metrics.txt --REMOVE_DUPLICATES true --MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP 50000 --MAX_FILE_HANDLES_FOR_READ_ENDS_MAP 8000 --SORTING_COLLECTION_SIZE_RATIO 0.25 --TAG_DUPLICATE_SET_MEMBERS false --REMOVE_SEQUENCING_DUPLICATES false --TAGGING_POLICY DontTag --CLEAR_DT true --DUPLEX_UMI false --FLOW_MODE false --FLOW_QUALITY_SUM_STRATEGY false --USE_END_IN_UNPAIRED_READS false --USE_UNPAIRED_CLIPPED_END false --UNPAIRED_END_UNCERTAINTY 0 --FLOW_SKIP_FIRST_N_FLOWS 0 --FLOW_Q_IS_KNOWN_END false --FLOW_EFFECTIVE_QUALITY_THRESHOLD 15 --ADD_PG_TAG_TO_READS true --ASSUME_SORTED false --DUPLICATE_SCORING_STRATEGY SUM_OF_BASE_QUALITIES --PROGRAM_RECORD_ID MarkDuplicates --PROGRAM_GROUP_NAME MarkDuplicates --READ_NAME_REGEX <optimized capture of last three ':' separated fields as numeric values> --OPTICAL_DUPLICATE_PIXEL_DISTANCE 100 --MAX_OPTICAL_DUPLICATE_SET_SIZE 300000 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
            "[Fri Dec 12 12:39:48 UTC 2025] Executing as root@6d3c47095ffb on Linux 6.6.105+ amd64; OpenJDK 64-Bit Server VM 11.0.27-internal+0-adhoc..src; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: Version:4.3.0.0\n",
            "INFO\t2025-12-12 12:39:48\tMarkDuplicates\tStart of doWork freeMemory: 64050344; totalMemory: 91226112; maxMemory: 3401580544\n",
            "INFO\t2025-12-12 12:39:48\tMarkDuplicates\tReading input file and constructing read end information.\n",
            "INFO\t2025-12-12 12:39:48\tMarkDuplicates\tWill retain up to 12324567 data points before spilling to disk.\n",
            "WARNING\t2025-12-12 12:39:48\tAbstractOpticalDuplicateFinderCommandLineProgram\tA field field parsed out of a read name was expected to contain an integer and did not. Read name: SRR15369234.156650288. Cause: String 'SRR15369234.156650288' did not start with a parsable number.\n",
            "INFO\t2025-12-12 12:40:01\tMarkDuplicates\tRead     1,000,000 records.  Elapsed time: 00:00:12s.  Time for last 1,000,000:   12s.  Last read position: CM037654.1:132,343,268\n",
            "INFO\t2025-12-12 12:40:01\tMarkDuplicates\tTracking 2868 as yet unmatched pairs. 17 records in RAM.\n",
            "INFO\t2025-12-12 12:40:10\tMarkDuplicates\tRead 1966080 records. 0 pairs never matched.\n",
            "INFO\t2025-12-12 12:40:10\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 698199960; totalMemory: 1037041664; maxMemory: 3401580544\n",
            "INFO\t2025-12-12 12:40:10\tMarkDuplicates\tWill retain up to 106299392 duplicate indices before spilling to disk.\n",
            "INFO\t2025-12-12 12:40:11\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
            "INFO\t2025-12-12 12:40:12\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
            "INFO\t2025-12-12 12:40:12\tMarkDuplicates\tSorting list of duplicate records.\n",
            "INFO\t2025-12-12 12:40:12\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1242063560; totalMemory: 2121269248; maxMemory: 3401580544\n",
            "INFO\t2025-12-12 12:40:12\tMarkDuplicates\tMarking 1833 records as duplicates.\n",
            "INFO\t2025-12-12 12:40:12\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
            "INFO\t2025-12-12 12:40:12\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
            "INFO\t2025-12-12 12:40:31\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
            "INFO\t2025-12-12 12:40:31\tMarkDuplicates\tDuplicate Index cleanup.\n",
            "INFO\t2025-12-12 12:40:31\tMarkDuplicates\tGetting Memory Stats.\n",
            "INFO\t2025-12-12 12:40:31\tMarkDuplicates\tBefore output close freeMemory: 476972536; totalMemory: 2121269248; maxMemory: 3401580544\n",
            "INFO\t2025-12-12 12:40:31\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
            "INFO\t2025-12-12 12:40:31\tMarkDuplicates\tAfter output close freeMemory: 65782408; totalMemory: 94371840; maxMemory: 3401580544\n",
            "[Fri Dec 12 12:40:31 UTC 2025] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.73 minutes.\n",
            "Runtime.totalMemory()=94371840\n",
            "Using GATK jar /usr/local/miniconda/envs/workshop_ngs/share/gatk4-4.3.0.0-0/gatk-package-4.3.0.0-local.jar\n",
            "Running:\n",
            "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /usr/local/miniconda/envs/workshop_ngs/share/gatk4-4.3.0.0-0/gatk-package-4.3.0.0-local.jar MarkDuplicates -I mapped_reads/BEN_CI16.bam -O mapped_reads/BEN_CI16.deduplicated.bam -M mapped_reads/BEN_CI16.metrics.txt --REMOVE_DUPLICATES true\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libtinfow.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n",
            "samtools: /usr/local/miniconda/envs/workshop_ngs/bin/../lib/libncursesw.so.6: no version information available (required by samtools)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "ls -ltrh mapped_reads/"
      ],
      "metadata": {
        "id": "0mfnwmFpnajB",
        "outputId": "bf4bd9cf-0214-4521-a438-cc9e590ea9ed",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total 478M\n",
            "-rw-r--r-- 1 root root 203M Dec 12 12:37 BEN_CI16.bam\n",
            "-rw-r--r-- 1 root root 273M Dec 12 12:40 BEN_CI16.deduplicated.bam\n",
            "-rw-r--r-- 1 root root 3.7K Dec 12 12:40 BEN_CI16.metrics.txt\n",
            "-rw-r--r-- 1 root root 3.3M Dec 12 12:40 BEN_CI16.deduplicated.bam.bai\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Tutorial Exercise\n",
        "\n",
        "Once you have performed trimming and mapping and completed till indexing steps given above, same can be done for the other two fastq files which we have downloaded (BEN_NW10 and BEN_SI18).\n",
        "\n",
        "Steps for the same are given below."
      ],
      "metadata": {
        "id": "sOiWTqu5o5n-"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Trimming**"
      ],
      "metadata": {
        "id": "Ik1RVN1Wqt9D"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "import os\n",
        "\n",
        "samples = [\"BEN_NW10\", \"BEN_SI18\"]\n",
        "output_dir = \"trimmed_fastq_files\"\n",
        "os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "for prefix in samples:\n",
        "    read1 = f\"fastq_files/{prefix}_sub_1.fq.gz\"\n",
        "    read2 = f\"fastq_files/{prefix}_sub_2.fq.gz\"\n",
        "\n",
        "    print(f\"\\n--- Running Trim Galore for {prefix} ---\")\n",
        "\n",
        "    # Run Trim Galore inside conda environment\n",
        "    !bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        "              conda activate workshop_ngs && \\\n",
        "              trim_galore --paired {read1} {read2} -o {output_dir}\"\n",
        "\n",
        "    # Remove extra report files\n",
        "    !rm -f {output_dir}/{prefix}_sub_*.fq.gz_trimming_report.txt\n",
        "    !rm -f {output_dir}/{prefix}_sub_*_fastqc.html\n",
        "    !rm -f {output_dir}/{prefix}_sub_*_fastqc.zip\n",
        "\n",
        "print(\"\\n--- Trimmed files ---\")\n",
        "!ls -F trimmed_fastq_files/\n"
      ],
      "metadata": {
        "id": "IUr6WXTh7zVy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Mapping**"
      ],
      "metadata": {
        "id": "gwOLBLtAqwWY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "REFERENCE_GENOME = \"reference/GCA_021130815.1_PanTigT.MC.v3_genomic.fna\"\n",
        "os.makedirs(\"mapped_reads\", exist_ok=True)\n",
        "\n",
        "samples = [\"BEN_NW10\", \"BEN_SI18\"]\n",
        "\n",
        "for prefix in samples:\n",
        "\n",
        "    read1 = f\"trimmed_fastq_files/{prefix}_sub_1_val_1.fq.gz\"\n",
        "    read2 = f\"trimmed_fastq_files/{prefix}_sub_2_val_2.fq.gz\"\n",
        "    output_bam = f\"mapped_reads/{prefix}.bam\"\n",
        "\n",
        "    print(f\"\\n--- Mapping {prefix} ---\")\n",
        "\n",
        "    # Map  convert  sort\n",
        "    !bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        "              conda activate workshop_ngs && \\\n",
        "              bwa mem -M -R '@RG\\tID:{prefix}\\tSM:{prefix}\\tPL:ILLUMINA' \\\n",
        "              {REFERENCE_GENOME} {read1} {read2} | \\\n",
        "              samtools view -bS - | \\\n",
        "              samtools sort -o {output_bam} -\"\n",
        "\n",
        "    # Index BAM\n",
        "   # print(f\"Indexing BAM for {prefix}...\")\n",
        "  #  !bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        "        #      conda activate workshop_ngs && \\\n",
        "      #        samtools index {output_bam}\"\n",
        "\n",
        "\n",
        "!ls -F mapped_reads/\n"
      ],
      "metadata": {
        "id": "nUI-4nLp71fe"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Mark Duplicates + Indexing**"
      ],
      "metadata": {
        "id": "INA7a0DhqyV8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "samples = [\"BEN_NW10\", \"BEN_SI18\"]\n",
        "\n",
        "for prefix in samples:\n",
        "    print(f\"\\n--- Running MarkDuplicates for {prefix} ---\")\n",
        "\n",
        "    !bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        "              conda activate workshop_ngs && \\\n",
        "              gatk MarkDuplicates \\\n",
        "                -I mapped_reads/{prefix}.sorted.bam \\\n",
        "                -O mapped_reads/{prefix}.deduplicated.bam \\\n",
        "                -M mapped_reads/{prefix}.metrics.txt \\\n",
        "                --REMOVE_DUPLICATES true\"\n",
        "\n",
        "    print(f\"\\n--- Indexing deduplicated BAM for {prefix} ---\")\n",
        "\n",
        "    !bash -c \"source /usr/local/miniconda/etc/profile.d/conda.sh && \\\n",
        "              conda activate workshop_ngs && \\\n",
        "              samtools index mapped_reads/{prefix}.deduplicated.bam\"\n",
        "\n",
        "print(\"\\n--- Finished MarkDuplicates + Indexing for all samples ---\")\n",
        "!ls -lh mapped_reads/\n"
      ],
      "metadata": {
        "id": "XXNUpi7Cqb5_"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}