{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/ternithinator/test2/blob/main/NGS_collab_snp_identification.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Variant calling is the process of identifying genetic variants, such as single nucleotide polymorphisms (SNPs) and small insertions or deletions (indels) from next-generation sequencing (NGS) data. It involves comparing sequencing reads from an individual to a reference genome to detect differences, like single nucleotide variants (SNVs). In this tutorial, we will focus specifically on detecting SNPs and indels.\n",
        "\n",
        "In this step:\n",
        "\n",
        "a) Call variants (Single Nucleotide Polymorphisms and Indels) using tools like bcftools or strelka\n",
        "\n",
        "b) Produce a VCF file (Variant Call Format) with detailed information about each variant\n",
        "\n",
        "c) Optionally apply variant filtering to remove false positives or low-confidence variants\n",
        "\n",
        "Why it's important: This is the core step of population genetics—identifying the genetic differences across sample.\n"
      ],
      "metadata": {
        "id": "nDRQBoNOVaIn"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "For this tutorial we are going to use Strelka, a tool utilised for germline and somatic variant calling.\n",
        "\n",
        "1: Germline Calling: Utilizes haplotype-based model to accurately detect inherited variants.\n",
        "\n",
        "(Haplotype is a set of DNA variants inherited together on the same chromosome copy)\n",
        "\n",
        "2: Somatic Calling: identifying genetic mutations that arise in somatic (non-germline) cells. These mutations are not inherited from parents and do not get passed on to offspring.\n",
        "\n",
        "Workflow Execution: Strelka2 can be run in two steps: configuration (specifying input data) and execution (specifying parameters)."
      ],
      "metadata": {
        "id": "8nEyHyy0Vn9A"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f3cc0067"
      },
      "source": [
        "\n",
        "The first step is to download the Miniconda installer for Linux using `wget`and create an environment for installing Srelka + Samtools."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "f6704e35",
        "collapsed": true
      },
      "source": [
        "%%bash\n",
        "# Download and install Miniconda\n",
        "wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh\n",
        "bash miniconda.sh -b -p /usr/local/miniconda\n",
        "\n",
        "# Add conda to PATH and initialize shell\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "# Accept Terms of Service for required conda channels\n",
        "conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/main\n",
        "conda tos accept --override-channels --channel https://repo.anaconda.com/pkgs/r\n",
        "\n",
        "# Create environment and install Strelka + Samtools\n",
        "conda create -y -n strelka_samtools -c bioconda strelka samtools"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Explanation**\n",
        "\n",
        "#### 1) Download Miniconda installer\n",
        "\n",
        "```bash\n",
        "wget -q https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh\n",
        "```\n",
        "\n",
        "* Downloads the latest Miniconda installer for Linux.\n",
        "* `-q` → quiet mode.\n",
        "* Saves file as `miniconda.sh`.\n",
        "\n",
        "#### 2) Install Miniconda silently\n",
        "\n",
        "```bash\n",
        "bash miniconda.sh -b -p /usr/local/miniconda\n",
        "```\n",
        "* Runs installer in **batch mode** (`-b`) → no user interaction needed.\n",
        "* Installs conda into `/usr/local/miniconda`.\n",
        "\n",
        "\n",
        "#### 3) Add conda to PATH + initialize shell\n",
        "\n",
        "```bash\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "```\n",
        "\n",
        "* Makes the `conda` command visible to your shell.\n",
        "* Loads the conda environment activation function.\n",
        "\n",
        "#### 4) Create a new environment with Strelka + Samtools\n",
        "\n",
        "```bash\n",
        "conda create -y -n strelka_samtools -c bioconda strelka samtools\n",
        "```\n",
        "\n",
        "* `-y` → auto-confirm installation.\n",
        "* `-n strelka` → environment will be named **strelka**.\n",
        "* `-c bioconda` → install packages from **Bioconda**, which hosts bioinformatics tools.\n",
        "* Installs both:\n",
        "\n",
        "  * **Strelka** → variant caller for germline & somatic variants\n",
        "  * **Samtools** → essential tool for BAM/SAM handling\n",
        "---"
      ],
      "metadata": {
        "id": "KToO07OiNgII"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "List packages inside the environment `strelka_samtools`"
      ],
      "metadata": {
        "id": "Cl4CLfV0Olla"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "\n",
        "conda list -n strelka_samtools"
      ],
      "metadata": {
        "id": "tqnV4USff2Hh",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "We will run this step `which configureStrelkaGermlineWorkflow.py` only to check whether Strelka is installed correctly and available in our PATH."
      ],
      "metadata": {
        "id": "41r7Kk_gDRvu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate strelka_samtools\n",
        "\n",
        "which configureStrelkaGermlineWorkflow.py"
      ],
      "metadata": {
        "id": "uALeX4fFDiYd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "After this we will create a new directory `input_files` in which reference genome `FNA` file will be downloaded from the zenodo repository."
      ],
      "metadata": {
        "id": "pA0iLoYDQeCn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "rm -rf input_files\n",
        "\n",
        "# Create the directory if it doesn't exist\n",
        "mkdir -p input_files\n",
        "\n",
        "# URL for reference genome tar.gz file\n",
        "wget -P input_files  https://zenodo.org/records/14258052/files/GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
        "\n",
        "# List extracted contents\n",
        "ls -F input_files/"
      ],
      "metadata": {
        "id": "t72HE-onDRIf",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next, we will generate **FASTA index file** of the reference file using Samtools. So that Strelka can quickly find any postiion in the genome without scanning the entire file."
      ],
      "metadata": {
        "id": "9SUB-qF4-pGF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate conda environment that contains samtools\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate strelka_samtools\n",
        "\n",
        "# Move into input_files directory\n",
        "cd input_files\n",
        "\n",
        "# Index the reference FASTA\n",
        "samtools faidx GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
        "\n",
        "#List the file\n",
        "ls -F\n"
      ],
      "metadata": {
        "id": "0hO1HqtS_V9p",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Explanation**\n",
        "\n",
        "`samtools faidx` is a command that creates a FASTA index (`.fai` file), which stores the exact byte positions of each chromosome so genomic regions can be accessed instantly.\n",
        "\n",
        "---\n"
      ],
      "metadata": {
        "id": "fP7IrfZzTv07"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "For the variant calling step, we will download the BAM file which is aligned, sorted, and duplicate-marked file (generated during the mapping step), from the Zenodo repository. This will ensure that they are ready for use in downstream variant identification."
      ],
      "metadata": {
        "id": "au3HYKKI_WMP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "# Download all BAM and BAI files into bam_files/\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_CI16_aligned_reads_deduplicated.bam?download=1\"     -O input_files/BEN_CI16_aligned_reads_sorted_deduplicated.bam\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_CI16_aligned_reads_deduplicated.bam.bai?download=1\" -O input_files/BEN_CI16_aligned_reads_sorted_deduplicated.bam.bai\n",
        "\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_NW10_aligned_reads_deduplicated.bam?download=1\"     -O input_files/BEN_NW10_aligned_reads_sorted_deduplicated.bam\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_NW10_aligned_reads_deduplicated.bam.bai?download=1\" -O input_files/BEN_NW10_aligned_reads_sorted_deduplicated.bam.bai\n",
        "\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_SI18_aligned_reads_deduplicated.bam?download=1\"     -O input_files/BEN_SI18_aligned_reads_sorted_deduplicated.bam\n",
        "wget -q \"https://zenodo.org/records/17895817/files/BEN_SI18_aligned_reads_deduplicated.bam.bai?download=1\" -O input_files/BEN_SI18_aligned_reads_sorted_deduplicated.bam.bai\n",
        "\n",
        "# List downloaded files\n",
        "ls -lh input_files\n"
      ],
      "metadata": {
        "id": "ejHSeGlmLNvu",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "Once we have our required BAM files, we will start the step for variant calling. In this analysis we use the Strelka germline workflow, which is initiated through the `configureStrelkaGermlineWorkflow.py` script.\n",
        "\n",
        "This germline configuration file contains:\n",
        "\n",
        "1: Information about the input BAM files (aligned, sorted, and duplicate-marked reads).\n",
        "\n",
        "2: The reference genome path, which Strelka uses for aligning sequences and identifying variant sites.\n",
        "\n",
        "3: Workflow parameters, such as filtering settings, runtime options, and rules for how Strelka processes sequencing data.\n",
        "\n",
        "4: Module definitions, specifying the order and structure of the processing steps used to detect SNVs and small indels.\n",
        "\n",
        "5: Paths to output directories, where Strelka will write variant call results, logs, and intermediate files.\n",
        "\n",
        "In short, this germline configuration is used by Strelka to build and execute the variant calling workflow. It ensures that the pipeline runs consistently with the correct inputs, reference genome, and settings needed for accurate germline variant detection.\\"
      ],
      "metadata": {
        "id": "XyaYBoxiLNH3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate strelka_samtools\n",
        "\n",
        "# Create the run directory manually (recommended for clarity)\n",
        "mkdir -p strelka_run\n",
        "\n",
        "# Configure Strelka workflow\n",
        "configureStrelkaGermlineWorkflow.py \\\n",
        "    --bam input_files/BEN_CI16_aligned_reads_sorted_deduplicated.bam \\\n",
        "    --bam input_files/BEN_NW10_aligned_reads_sorted_deduplicated.bam \\\n",
        "    --bam input_files/BEN_SI18_aligned_reads_sorted_deduplicated.bam \\\n",
        "    --referenceFasta input_files/GCA_021130815.1_PanTigT.MC.v3_genomic.fna \\\n",
        "    --runDir strelka_run"
      ],
      "metadata": {
        "id": "vsYRDIa59Bj3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "###**Explanation**\n",
        "\n",
        "#### 1) Create a run directory\n",
        "\n",
        "```bash\n",
        "mkdir -p strelka_run\n",
        "```\n",
        "\n",
        "\n",
        "Strelka needs a clean directory to store: workflow scripts, configuration files, logs and final outputs.\n",
        "\n",
        "\n",
        "#### 2) Configure the Strelka Germline Workflow\n",
        "\n",
        "```bash\n",
        "configureStrelkaGermlineWorkflow.py \\\n",
        "    --bam input_files/BEN_CI16_aligned_reads_sorted_deduplicated.bam \\\n",
        "    --bam input_files/BEN_NW10_aligned_reads_sorted_deduplicated.bam \\\n",
        "    --bam input_files/BEN_SI18_aligned_reads_sorted_deduplicated.bam \\\n",
        "    --referenceFasta input_files/GCA_021130815.1_PanTigT.MC.v3_genomic.fna \\\n",
        "    --runDir strelka_run\n",
        "```\n",
        "\n",
        "`configureStrelkaGermlineWorkflow.py` This is the Strelka script that **creates the entire workflow**, including makefiles, configuration JSON files, and scripts to run variant calling.\n",
        "\n",
        "\n",
        "```bash\n",
        "--bam input_files/BEN_CI16_aligned_reads_sorted_deduplicated.bam\n",
        "--bam input_files/BEN_NW10_aligned_reads_sorted_deduplicated.bam\n",
        "--bam input_files/BEN_SI18_aligned_reads_sorted_deduplicated.bam\n",
        "```\n",
        "\n",
        "Provides all BAM files\n",
        "\n",
        "\n",
        "```bash\n",
        "--referenceFasta input_files/GCA_021130815.1_PanTigT.MC.v3_genomic.fna\n",
        "```\n",
        "\n",
        "Tells Strelka **which genome assembly** was used to align the reads.\n",
        "\n",
        "\n",
        "```bash\n",
        "--runDir strelka_run\n",
        "```\n",
        "Instructs Strelka to create the full workflow inside the folder `strelka_run`.\n",
        "\n",
        "---"
      ],
      "metadata": {
        "id": "bM46i9lIVGJ1"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "In the final step, Strelka enters the workflow that was previously created inside the `strelka_run/` directory and executes all steps of the germline variant-calling pipeline. It automatically processes all the BAM files. It performs variant calling on each sample using the reference genome, and then generates the final SNP and INDEL output files inside `strelka_run/results/`.\n",
        "\n"
      ],
      "metadata": {
        "id": "qUOeujkhXyEN"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Activate environment\n",
        "export PATH=\"/usr/local/miniconda/bin:$PATH\"\n",
        "source /usr/local/miniconda/etc/profile.d/conda.sh\n",
        "conda activate strelka_samtools\n",
        "\n",
        "strelka_run/runWorkflow.py -m local -j 4"
      ],
      "metadata": {
        "id": "HQwRwuGNF-hz",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "###**Explanation**\n",
        "* `-m local` Runs Strelka on local machine.\n",
        "\n",
        "* `-j 4` Uses 4 CPU threads in parallel to speed up variant calling.\n",
        "\n"
      ],
      "metadata": {
        "id": "2rBxMoxKXDg8"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "\n",
        "After running the Strelka germline variant calling workflow, the pipeline produces a set of output files, most importantly the **VCF (Variant Call Format) files**. These files contain the list of single nucleotide variants (SNVs) and small insertions/deletions (indels) identified from the input BAM files.\n",
        "\n",
        "Once the VCF files are generated, they represent the final variant calls produced by Strelka, and they can now be used for downstream analyses. Typical post-processing steps may include:\n",
        "\n",
        "* **Quality assessment** of the variants by inspecting Strelka’s FILTER and quality score annotations.\n",
        "* **Merging or comparing VCFs** (if multiple samples were processed) to evaluate shared or unique variants.\n",
        "* **Functional annotation** using external tools (e.g., VEP, SnpEff) to predict gene impacts, consequences, or biological relevance.\n",
        "* **Filtering based on depth, genotype quality, or allele frequency**, depending on the research goals.\n",
        "\n",
        "In summary, the generation of VCF files marks the completion of the variant calling workflow, providing a structured dataset of all detected variants. These files form the foundation for interpretation, annotation, and further biological or population-level analyses.\n"
      ],
      "metadata": {
        "id": "cJ8qzZir9B7r"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Tutorial Questions\n",
        "\n",
        "1) What basic information does each row in a VCF file represent (e.g. position, reference base, alternative base)?\n",
        "\n",
        "2) Where in the VCF can you see whether a variant is homozygous or heterozygous?\n",
        "\n",
        "3) What does PASS mean in the FILTER column of a Strelka VCF?\n",
        "\n",
        "4) Why do you think Strelka produces separate VCF files for SNPs and indels?\n",
        "\n",
        "5) Which column in the VCF tells you how confident the variant call is, and what happens to this value when the evidence for a variant is weak?"
      ],
      "metadata": {
        "id": "IKzyMF-QNFlX"
      }
    }
  ]
}